{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       ".container {\n",
       "    width: 100%;\n",
       "}\n",
       "div#notebook {\n",
       "    padding-top: 1em;\n",
       "}\n",
       "#header-container {\n",
       "    display: none;\n",
       "}\n",
       "#header-bar {\n",
       "    display: none;\n",
       "}\n",
       "#maintoolbar {\n",
       "    display: none;\n",
       "}\n",
       "#menubar-container {\n",
       "    position: fixed;\n",
       "    margin-top: 0;\n",
       "}\n",
       "#site {\n",
       "    height: auto !important;\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%run setup.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hierarchical clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function to support searching a scipy tree.\n",
    "\n",
    "def get_descendant(node, desc_id):\n",
    "    \"\"\"Search the descendants of the given node.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    node : scipy.cluster.hierarchy.ClusterNode\n",
    "        The ancestor node to search from.\n",
    "    desc_id : int\n",
    "        The ID of the node to search for.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    desc : scipy.cluster.hierarchy.ClusterNode\n",
    "        If a node with the given ID is not found, returns None.\n",
    "    \n",
    "    \"\"\"\n",
    "    if node.id == desc_id:\n",
    "        return node\n",
    "    if node.is_leaf():\n",
    "        return None\n",
    "    if node.left.id == desc_id:\n",
    "        return node.left\n",
    "    if node.right.id == desc_id:\n",
    "        return node.right\n",
    "    # search left\n",
    "    l = get_descendant(node.left, desc_id)\n",
    "    if l is not None:\n",
    "        return l\n",
    "    # search right\n",
    "    r = get_descendant(node.right, desc_id)\n",
    "    return r\n",
    "\n",
    "\n",
    "# monkey-patch as a method\n",
    "scipy.cluster.hierarchy.ClusterNode.get_descendant = get_descendant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fig_haplotypes_clustered(h,\n",
    "                             distance_metric='hamming',\n",
    "                             linkage_method='single',\n",
    "                             truncate_distance=0,\n",
    "                             orientation='top',\n",
    "                             subplot_ratios=(4, 2),\n",
    "                             subplot_pad=0,\n",
    "                             despine_offset=5,\n",
    "                             count_sort=True,\n",
    "                             dend_linecolor='k',\n",
    "                             cut_height=2,\n",
    "                             highlight_clusters=True,\n",
    "                             highlight_colors=None,\n",
    "                             highlight_dend=True,\n",
    "                             highlight_freq=True,\n",
    "                             highlight_alpha=0.3,\n",
    "                             label_clusters=True,\n",
    "                             dpi=None,\n",
    "                             fig=None,\n",
    "                             ):\n",
    "    \"\"\"Construct a plot of hierarchical clustering of haplotypes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    TODO\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    TODO\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # check inputs\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    \n",
    "    # compute distance matrix\n",
    "    dist = scipy.spatial.distance.pdist(h.T, metric=distance_metric)\n",
    "    if distance_metric in {'hamming', 'jaccard'}:\n",
    "        # convert distance to number of SNPs, easier to interpret\n",
    "        dist *= h.n_variants\n",
    "    \n",
    "    # compute hierarchical clustering\n",
    "    Z = scipy.cluster.hierarchy.linkage(dist, method=linkage_method)\n",
    "\n",
    "    # Z is a linkage matrix. From the scipy docs...\n",
    "    # A 4 by (n-1) matrix Z is returned. At the i-th iteration, clusters with \n",
    "    # indices Z[i, 0] and Z[i, 1] are combined to form cluster n + i. A cluster \n",
    "    # with an index less than n corresponds to one of the original observations. \n",
    "    # The distance between clusters Z[i, 0] and Z[i, 1] is given by Z[i, 2]. The \n",
    "    # fourth value Z[i, 3] represents the number of original observations in the \n",
    "    # newly formed cluster.\n",
    "    \n",
    "    # find level to truncate dendrogram\n",
    "    lastp = h.n_haplotypes - bisect.bisect_right(Z[:, 2], truncate_distance)\n",
    "    \n",
    "    # convenience variables\n",
    "    horizontal = orientation in ['left', 'right']\n",
    "    vertical = not horizontal\n",
    "    inverted = orientation in ['bottom', 'right']\n",
    "    \n",
    "    # setup figure\n",
    "    if fig is None:\n",
    "        figsize = plt.rcParams['figure.figsize']\n",
    "        if horizontal:\n",
    "            figsize = figsize[::-1]\n",
    "        fig = plt.figure(figsize=figsize, dpi=dpi)\n",
    "        \n",
    "    # setup gridspec and axes\n",
    "    if inverted:\n",
    "        subplot_ratios = subplot_ratios[::-1]\n",
    "    if horizontal:\n",
    "        gs = plt.GridSpec(nrows=1, ncols=2, width_ratios=subplot_ratios)\n",
    "    else:\n",
    "        gs = plt.GridSpec(nrows=2, ncols=1, height_ratios=subplot_ratios)\n",
    "    if inverted:\n",
    "        ax_dend = fig.add_subplot(gs[1])\n",
    "        ax_freq = fig.add_subplot(gs[0])\n",
    "    else:\n",
    "        ax_dend = fig.add_subplot(gs[0])\n",
    "        ax_freq = fig.add_subplot(gs[1])\n",
    "    if horizontal:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "    else:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "\n",
    "    # make a dendrogram\n",
    "    kwargs_dend = dict(\n",
    "        truncate_mode='lastp', \n",
    "        p=lastp,\n",
    "        show_leaf_counts=False, \n",
    "        count_sort=count_sort, \n",
    "        no_labels=True, \n",
    "        color_threshold=0, \n",
    "        above_threshold_color=dend_linecolor, \n",
    "        orientation=orientation\n",
    "    )    \n",
    "    dend = scipy.cluster.hierarchy.dendrogram(Z, ax=ax_dend, **kwargs_dend)\n",
    "    leaves = dend['leaves']\n",
    "    ax_dend_label = 'Distance'\n",
    "    if horizontal:\n",
    "        ax_dend.set_xlabel(ax_dend_label)\n",
    "        ax_dend.set_yticks([])\n",
    "    else:\n",
    "        ax_dend.set_ylabel(ax_dend_label)\n",
    "        ax_dend.set_xticks([])\n",
    "        \n",
    "    # construct a tree and compute observation counts for the dendrogram leaves\n",
    "    tree = scipy.cluster.hierarchy.to_tree(Z)\n",
    "    s = np.arange(len(leaves))\n",
    "    t = np.array([\n",
    "        1 if l < h.n_haplotypes\n",
    "        else tree.get_descendant(l).get_count()\n",
    "        for l in leaves\n",
    "    ])\n",
    "\n",
    "    # plot frequencies bar\n",
    "    ax_freq_label = 'Frequency'\n",
    "    if horizontal:\n",
    "        ax_freq.barh(s, t, height=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_ylim(0, len(leaves))\n",
    "        ax_freq.set_yticks([])\n",
    "        ax_freq.set_xlabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='x', lw=.5)\n",
    "        if orientation == 'right':\n",
    "            ax_freq.invert_xaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_xticks(ax_freq.get_xticks()[1:])\n",
    "    else:\n",
    "        ax_freq.bar(s, t, width=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_xlim(0, len(leaves))\n",
    "        ax_freq.set_xticks([])\n",
    "        ax_freq.set_ylabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='y', lw=.5)\n",
    "        if orientation == 'top':\n",
    "            ax_freq.invert_yaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_yticks(ax_freq.get_yticks()[1:])\n",
    "\n",
    "    # cut the tree\n",
    "    cut = scipy.cluster.hierarchy.cut_tree(Z, height=cut_height)[:, 0]\n",
    "    cluster_sizes = np.bincount(cut)\n",
    "    clusters = [np.nonzero(cut == i)[0] for i in range(cut.max() + 1)]\n",
    "    \n",
    "    # now the fiddly bit - we need to figure out where the clusters have\n",
    "    # ended up in the dendrogram we plotted earlier...\n",
    "    \n",
    "    # N.B., the dendrogram was truncated, so each leaf in the dendrogram\n",
    "    # may correspond to more than one original observation (i.e., haplotype).\n",
    "    # Let's build a list storing the observations for each leaf:\n",
    "    leaf_obs = [tree.get_descendant(ix).pre_order() for ix in leaves]\n",
    "    \n",
    "    # Now let's figure out for each leaf in the dendrogram, which of the clusters\n",
    "    # obtained by cutting the tree earlier does it fall into?\n",
    "    leaf_clusters = np.array([cut[l[0]] for l in leaf_obs])\n",
    "    \n",
    "    # Now let's build a data structure that reorders the clusters so they\n",
    "    # occur in the same order as in the dendrogram, and also record the indices\n",
    "    # of the start and stop leaf for each cluster:\n",
    "    cluster_spans = list()\n",
    "    c_prv = leaf_clusters[0]\n",
    "    i_start = 0\n",
    "    for i, c in enumerate(leaf_clusters[1:], 1):\n",
    "        if c != c_prv:\n",
    "            cluster_spans.append((i_start, i, clusters[c_prv]))\n",
    "            i_start = i\n",
    "        c_prv = c\n",
    "    # don't forget the last one\n",
    "    cluster_spans.append((i_start, i+1, clusters[c]))\n",
    "    \n",
    "    # OK, now figure out which clusters we want to highlight...\n",
    "    if isinstance(highlight_clusters, (list, tuple)):\n",
    "        # user has manually specified which clusters to highlight\n",
    "        pass\n",
    "    else:\n",
    "        # assume highlight_clusters is the minimum cluster size to highlight\n",
    "        min_cluster_size = int(highlight_clusters)\n",
    "        highlight_clusters = [i for i, cs in enumerate(cluster_spans) \n",
    "                              if len(cs[2]) >= min_cluster_size]\n",
    "        \n",
    "    # setup colors for highlighting clusters\n",
    "    if highlight_colors is None:\n",
    "        highlight_colors = sns.color_palette('hls', n_colors=len(highlight_clusters))\n",
    "        \n",
    "    # do the highlighting\n",
    "    for color, cix in zip(highlight_colors, highlight_clusters):\n",
    "        start, stop, _ = cluster_spans[cix]\n",
    "        if horizontal:\n",
    "            freq_spanf = ax_freq.axhspan\n",
    "            dend_patch_xy = (0, start * 10)\n",
    "            dend_patch_width = cut_height\n",
    "            dend_patch_height = (stop - start) * 10\n",
    "        else:\n",
    "            freq_spanf = ax_freq.axvspan\n",
    "            dend_patch_xy = (start * 10, 0)\n",
    "            dend_patch_width = (stop - start) * 10\n",
    "            dend_patch_height = cut_height\n",
    "        if highlight_freq:\n",
    "            freq_spanf(start, stop, color=color, alpha=highlight_alpha, zorder=-20)\n",
    "        if highlight_dend:\n",
    "            ax_dend.add_patch(plt.Rectangle(xy=dend_patch_xy, \n",
    "                                            width=dend_patch_width, \n",
    "                                            height=dend_patch_height, \n",
    "                                            color=color, alpha=highlight_alpha, \n",
    "                                            zorder=-20))\n",
    "\n",
    "    # for debugging, label the clusters\n",
    "    if label_clusters:\n",
    "        for i, (start, stop, clst) in enumerate(cluster_spans):\n",
    "            if horizontal:\n",
    "                x = max(ax_freq.get_xlim())\n",
    "                y = (start + stop) / 2\n",
    "                ha = orientation\n",
    "                va = 'center'\n",
    "            else:\n",
    "                x = (start + stop) / 2\n",
    "                y = max(ax_freq.get_ylim())\n",
    "                ha = 'center'\n",
    "                va = orientation\n",
    "            # treat label_clusters as minimum cluster size to label\n",
    "            if len(clst) >= int(label_clusters):\n",
    "                ax_freq.text(x, y, str(i), \n",
    "                             va=va, ha=ha, fontsize=6)\n",
    "    \n",
    "    # tidy up plot\n",
    "    if horizontal:\n",
    "        gs.tight_layout(fig, w_pad=subplot_pad)\n",
    "    else:\n",
    "        gs.tight_layout(fig, h_pad=subplot_pad)\n",
    "    \n",
    "    # return some useful stuff\n",
    "    return fig, ax_dend, ax_freq, cluster_spans, leaf_obs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotype networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _graph_edges(graph, \n",
    "                 edges, \n",
    "                 hap_counts, \n",
    "                 node_size_factor, \n",
    "                 edge_length,\n",
    "                 anon_width,\n",
    "                 intermediate_nodes,\n",
    "                 edge_attrs,\n",
    "                 anon_node_attrs):\n",
    "    \n",
    "    for i in range(edges.shape[0]):\n",
    "\n",
    "        for j in range(edges.shape[1]):\n",
    "\n",
    "            # lookup distance between nodes i and j\n",
    "            sep = edges[i, j]\n",
    "\n",
    "            if sep > 0:\n",
    "\n",
    "                # lookup number of haplotypes\n",
    "                # calculate node sizes (needed to adjust edge length)\n",
    "                if i < len(hap_counts):\n",
    "                    n_i = hap_counts[i]\n",
    "                    width_i = np.sqrt(n_i * node_size_factor)\n",
    "                else:\n",
    "                    n_i = 1\n",
    "                    width_i = anon_width\n",
    "                if j < len(hap_counts):\n",
    "                    n_j = hap_counts[j]\n",
    "                    width_j = np.sqrt(n_j * node_size_factor)\n",
    "                else:\n",
    "                    n_j = 1\n",
    "                    width_j = anon_width\n",
    "\n",
    "                if sep > 1 and intermediate_nodes:\n",
    "\n",
    "                    # tricky case, need to add some anonymous nodes to represent intermediate steps\n",
    "\n",
    "                    # add first intermediate node\n",
    "                    nid = 'anon_{}_{}_{}'.format(i, j, 0)\n",
    "                    graph.node(nid, label='', width=str(anon_width), **anon_node_attrs)\n",
    "\n",
    "                    # add edge from node i to first intermediate\n",
    "                    el = edge_length + width_i / 2 + anon_width / 2\n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge(str(i), 'anon_{}_{}_{}'.format(i, j, 0), **kwargs)\n",
    "\n",
    "                    # add further intermediate nodes as necessary\n",
    "                    for k in range(1, sep-1):\n",
    "                        nid = 'anon_{}_{}_{}'.format(i, j, k)\n",
    "                        graph.node(nid, label='', width=str(anon_width), **anon_node_attrs)\n",
    "                        el = edge_length + anon_width\n",
    "                        kwargs = {'len': str(el)}\n",
    "                        kwargs.update(edge_attrs)\n",
    "                        graph.edge('anon_{}_{}_{}'.format(i, j, k-1), 'anon_{}_{}_{}'.format(i, j, k), **kwargs)\n",
    "\n",
    "                    # add edge from final intermediate node to node j\n",
    "                    el = edge_length + anon_width / 2 + width_j / 2 \n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge('anon_{}_{}_{}'.format(i, j, sep-2), str(j), **kwargs)\n",
    "\n",
    "                else:\n",
    "\n",
    "                    # simple case, direct edge from node i to j\n",
    "\n",
    "                    # N.B., adjust edge length so we measure distance from edge of circle rather than center\n",
    "                    el = (edge_length * sep) + width_i / 2 + width_j / 2\n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge(str(i), str(j), **kwargs)\n",
    "\n",
    "                    \n",
    "def graph_haplotype_network(h,\n",
    "                            hap_colors='grey',\n",
    "                            distance_metric='hamming',\n",
    "                            network_method='mst',\n",
    "                            comment=None,\n",
    "                            engine='neato',\n",
    "                            format='png',\n",
    "                            mode='major',\n",
    "                            overlap=True,\n",
    "                            splines=False,\n",
    "                            graph_attrs=None,\n",
    "                            node_size_factor=0.01,\n",
    "                            node_attrs=None,\n",
    "                            show_node_labels=False,\n",
    "                            fontname='monospace',\n",
    "                            fontsize=None,\n",
    "                            edge_length=0.5,\n",
    "                            edge_weight=10,\n",
    "                            edge_attrs=None,\n",
    "                            show_alternate_edges=True,\n",
    "                            alternate_edge_attrs=None,\n",
    "                            anon_width=0.07,\n",
    "                            anon_fillcolor='white',\n",
    "                            anon_node_attrs=None,\n",
    "                            intermediate_nodes=True,\n",
    "                            max_dist=5,\n",
    "                            debug=False,\n",
    "                            ):\n",
    "    \"\"\"TODO doc me\"\"\"\n",
    "    \n",
    "    # check inputs\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    \n",
    "    # optimise - keep only segregating variants\n",
    "    ac = h.count_alleles()\n",
    "    h = h[ac.is_segregating()]\n",
    "    \n",
    "    # find distinct haplotypes\n",
    "    h_distinct_sets = h.distinct()\n",
    "    \n",
    "    # find indices of distinct haplotypes - just need one per set\n",
    "    h_distinct_indices = [sorted(s)[0] for s in h_distinct_sets]\n",
    "    \n",
    "    # reorder by index\n",
    "    ix = np.argsort(h_distinct_indices)\n",
    "    h_distinct_indices = [h_distinct_indices[i] for i in ix]\n",
    "    h_distinct_sets = [h_distinct_sets[i] for i in ix]\n",
    "    \n",
    "    # obtain an array of distinct haplotypes\n",
    "    h_distinct = h.take(h_distinct_indices, axis=1)\n",
    "\n",
    "    # deal with colors - count how many of each color per distinct haplotype\n",
    "    color_counters = None\n",
    "    if isinstance(hap_colors, (list, tuple, np.ndarray)):\n",
    "        assert len(hap_colors) == h.n_haplotypes\n",
    "        color_counters = [\n",
    "            collections.Counter([hap_colors[i] for i in s])\n",
    "            for s in h_distinct_sets\n",
    "        ]\n",
    "\n",
    "    # count how many observations per distinct haplotype\n",
    "    hap_counts = [len(s) for s in h_distinct_sets]\n",
    "\n",
    "    # compute pairwise distance matrix\n",
    "    assert distance_metric in ['hamming', 'jaccard']\n",
    "    dist = allel.pairwise_distance(h_distinct, metric=distance_metric)\n",
    "    dist *= h.n_variants\n",
    "    dist = scipy.spatial.distance.squareform(dist).astype(int)\n",
    "\n",
    "    if network_method.lower() == 'mst':\n",
    "\n",
    "        # compute minimum spanning tree\n",
    "        edges = scipy.sparse.csgraph.minimum_spanning_tree(dist).toarray().astype(int)\n",
    "\n",
    "        # deal with maximum distance\n",
    "        if max_dist:\n",
    "            edges[edges > max_dist] = 0\n",
    "\n",
    "        # no alternate edges when using mst\n",
    "        alternate_edges = None\n",
    "        \n",
    "    elif network_method.lower() == 'msn':\n",
    "        \n",
    "        # compute network\n",
    "        edges, alternate_edges = minimum_spanning_network(dist, max_dist=max_dist, debug=debug)\n",
    "        edges = np.triu(edges)\n",
    "        alternate_edges = np.triu(alternate_edges)\n",
    "        \n",
    "    elif network_method.lower() == 'mjn':\n",
    "        \n",
    "        # compute network\n",
    "        _, edges, alternate_edges = median_joining_network(h_distinct, max_dist=max_dist, debug=debug)\n",
    "        edges = np.triu(edges)\n",
    "        alternate_edges = np.triu(alternate_edges)\n",
    "        \n",
    "    else:\n",
    "        raise ValueError(network_method)\n",
    "\n",
    "    # setup graph\n",
    "    graph = graphviz.Graph(comment=comment, engine=engine, format=format)\n",
    "    if graph_attrs is None:\n",
    "        graph_attrs = dict()\n",
    "    graph_attrs.setdefault('overlap', str(overlap).lower())\n",
    "    graph_attrs.setdefault('splines', str(splines).lower())\n",
    "    graph_attrs.setdefault('mode', mode)\n",
    "    graph.attr('graph', **graph_attrs)\n",
    "\n",
    "    # add the main nodes\n",
    "    if node_attrs is None:\n",
    "        node_attrs = dict()\n",
    "    node_attrs.setdefault('fixedsize', 'true')\n",
    "    node_attrs.setdefault('shape', 'circle')\n",
    "    node_attrs.setdefault('fontname', fontname)\n",
    "    node_attrs.setdefault('fontsize', str(fontsize))\n",
    "    if anon_node_attrs is None:\n",
    "        anon_node_attrs = dict()\n",
    "    anon_node_attrs.setdefault('fixedsize', 'true')\n",
    "    anon_node_attrs.setdefault('shape', 'circle')\n",
    "    anon_node_attrs.setdefault('style', 'filled')\n",
    "    anon_node_attrs.setdefault('fillcolor', anon_fillcolor)\n",
    "    anon_node_attrs.setdefault('fontname', fontname)\n",
    "    anon_node_attrs.setdefault('fontsize', str(fontsize))\n",
    "    for i in range(edges.shape[0]):\n",
    "        kwargs = dict()\n",
    "        \n",
    "        if i < len(hap_counts):\n",
    "            n = hap_counts[i]\n",
    "\n",
    "            # calculate width from number of items - make width proportional to area\n",
    "            width = np.sqrt(n * node_size_factor)\n",
    "\n",
    "            # determine style and fill color\n",
    "            if color_counters:\n",
    "                cc = color_counters[i]\n",
    "                if len(cc) > 1:\n",
    "                    # more than one color, make a pie chart\n",
    "                    style = 'wedged'\n",
    "                    fillcolor = ':'.join(['%s;%s' % (k, v/n) for k, v in cc.items()])\n",
    "                else:\n",
    "                    # just one color, fill with solid color\n",
    "                    style = 'filled'\n",
    "                    fillcolor = list(cc.keys())[0]\n",
    "            else:\n",
    "                style = 'filled'\n",
    "                fillcolor = hap_colors\n",
    "            \n",
    "            kwargs.update(node_attrs)\n",
    "            kwargs.setdefault('style', style)\n",
    "            kwargs.setdefault('fillcolor', fillcolor)\n",
    "            kwargs.setdefault('width', str(width))\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            width = anon_width\n",
    "            fillcolor = anon_fillcolor\n",
    "            kwargs.update(anon_node_attrs)\n",
    "            kwargs.setdefault('width', str(anon_width))\n",
    "\n",
    "        # add graph node\n",
    "        if show_node_labels:\n",
    "            label = str(i)\n",
    "        else:\n",
    "            label = \"\"\n",
    "        kwargs.setdefault('label', label)\n",
    "        graph.node(str(i), **kwargs)\n",
    "    \n",
    "    # setup defaults\n",
    "    if edge_attrs is None:\n",
    "        edge_attrs = dict()\n",
    "    edge_attrs.setdefault('style', 'normal')\n",
    "    edge_attrs.setdefault('weight', str(edge_weight))\n",
    "    edge_attrs.setdefault('fontname', fontname)\n",
    "    edge_attrs.setdefault('fontsize', str(fontsize))\n",
    "    if alternate_edge_attrs is None:\n",
    "        alternate_edge_attrs = dict()\n",
    "    alternate_edge_attrs.setdefault('style', 'dashed')\n",
    "    alternate_edge_attrs.setdefault('weight', str(edge_weight))\n",
    "    alternate_edge_attrs.setdefault('fontname', fontname)\n",
    "    alternate_edge_attrs.setdefault('fontsize', str(fontsize))\n",
    "\n",
    "    # add main edges\n",
    "    _graph_edges(graph, \n",
    "                 edges, \n",
    "                 hap_counts, \n",
    "                 node_size_factor, \n",
    "                 edge_length,\n",
    "                 anon_width,\n",
    "                 intermediate_nodes,\n",
    "                 edge_attrs,\n",
    "                 anon_node_attrs)\n",
    "    \n",
    "    # add alternate edges\n",
    "    if show_alternate_edges and alternate_edges is not None:\n",
    "        _graph_edges(graph, \n",
    "                     alternate_edges, \n",
    "                     hap_counts, \n",
    "                     node_size_factor, \n",
    "                     edge_length,\n",
    "                     anon_width,\n",
    "                     intermediate_nodes,\n",
    "                     alternate_edge_attrs,\n",
    "                     anon_node_attrs)\n",
    "\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def minimum_spanning_network(dist, max_dist=None, debug=False):\n",
    "    \"\"\"TODO\"\"\"\n",
    "    \n",
    "    # TODO review implementation, see if this can be tidied up\n",
    "    \n",
    "    # keep only the lower triangle of the distance matrix, to avoid adding the same\n",
    "    # edge twice\n",
    "    dist = np.triu(dist)\n",
    "    \n",
    "    # setup the output array of links between nodes\n",
    "    edges = np.zeros_like(dist)\n",
    "    \n",
    "    # setup an array of alternate links\n",
    "    alternate_edges = np.zeros_like(dist)\n",
    "    \n",
    "    # intermediate variable - assignment of haplotypes to clusters (a.k.a. sub-networks)\n",
    "    # initially each distinct haplotype is in its own cluster\n",
    "    cluster = np.arange(dist.shape[0])\n",
    "    \n",
    "    # start with haplotypes separated by a single mutation\n",
    "    step = 1\n",
    "    \n",
    "    # iterate until all haplotypes in a single cluster, or max_dist reached\n",
    "    while len(set(cluster)) > 1 and (max_dist is None or step <= max_dist):\n",
    "        if debug: print('processing step', step, cluster)\n",
    "            \n",
    "        # keep track of which clusters have been merged at this height\n",
    "        merged = set()\n",
    "        \n",
    "        # remember what cluster assignments were at the previous height\n",
    "        prv_cluster = cluster.copy()\n",
    "        \n",
    "        # iterate over all pairs where distance equals current step size\n",
    "        for i, j in zip(*np.nonzero(dist == step)):\n",
    "            if debug: print('found potential edge', i, j)\n",
    "            \n",
    "            # current cluster assignment for each haplotype\n",
    "            a = cluster[i]\n",
    "            b = cluster[j]\n",
    "            \n",
    "            # previous cluster assignment for each haplotype\n",
    "            pa = prv_cluster[i]\n",
    "            pb = prv_cluster[j]\n",
    "            \n",
    "            if debug: print(a, b, pa, pb, merged)\n",
    "            \n",
    "            # check to see if both nodes already in the same cluster\n",
    "            if a != b:\n",
    "                \n",
    "                # nodes are in different clusters, so we can merge (i.e., connect) the clusters\n",
    "                \n",
    "                if debug: print('assign an edge')\n",
    "                edges[i, j] = dist[i, j]\n",
    "                edges[j, i] = dist[i, j]\n",
    "                \n",
    "                # merge clusters\n",
    "                c = cluster.max() + 1\n",
    "                loc_a = cluster == a\n",
    "                loc_b = cluster == b\n",
    "                cluster[loc_a] = c\n",
    "                cluster[loc_b] = c\n",
    "                merged.add(tuple(sorted([pa, pb])))\n",
    "                if debug: print('merged', cluster, merged)\n",
    "\n",
    "            elif tuple(sorted([pa, pb])) in merged or step == 1:\n",
    "                \n",
    "                # the two clusters have already been merged at this level, this is an alternate connection\n",
    "                # N.B., special case step = 1 because no previous cluster assignments\n",
    "                \n",
    "                if debug: print('assign an alternate edge')\n",
    "                alternate_edges[i, j] = dist[i, j]\n",
    "                alternate_edges[j, i] = dist[i, j]\n",
    "                \n",
    "            else:\n",
    "                \n",
    "                if debug: print('')\n",
    "                    \n",
    "        # increment step\n",
    "        step += 1\n",
    "                \n",
    "    return edges, alternate_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _remove_obsolete(h, orig_n_haplotypes, max_dist=None, debug=False):\n",
    "    n_removed = None\n",
    "    \n",
    "    while n_removed is None or n_removed > 0:\n",
    "\n",
    "        # step 1 - compute distance\n",
    "        dist = allel.pairwise_distance(h, metric='hamming')\n",
    "        dist *= h.n_variants\n",
    "        dist = scipy.spatial.distance.squareform(dist).astype(int)\n",
    "\n",
    "        # step 2 - construct the minimum spanning network\n",
    "        edges, alt_edges = minimum_spanning_network(dist, max_dist=max_dist)\n",
    "        all_edges = edges + alt_edges\n",
    "\n",
    "        # step 3 - remove obsolete sequence types\n",
    "        loc_keep = np.ones(h.n_haplotypes, dtype=bool)\n",
    "        for i in range(orig_n_haplotypes, h.n_haplotypes):\n",
    "            n_connections = np.count_nonzero(all_edges[i])\n",
    "            if n_connections <= 2:\n",
    "                loc_keep[i] = False\n",
    "        n_removed = np.count_nonzero(~loc_keep)\n",
    "        if debug: print('discarding', n_removed, 'obsolete haplotypes')\n",
    "        h = h[:, loc_keep]\n",
    "        \n",
    "    return h, edges, alt_edges\n",
    "    \n",
    "\n",
    "def median_joining_network(h, debug=False, max_dist=None):\n",
    "    h = allel.HaplotypeArray(h, dtype='i1')\n",
    "    orig_n_haplotypes = h.n_haplotypes\n",
    "    \n",
    "    n_medians_added = None\n",
    "    iteration = 0\n",
    "    while n_medians_added is None or n_medians_added > 0:\n",
    "\n",
    "        # steps 1-3\n",
    "        h, edges, alt_edges = _remove_obsolete(h, orig_n_haplotypes, max_dist=max_dist, debug=debug)\n",
    "        all_edges = edges + alt_edges\n",
    "\n",
    "        # step 4 - add median vectors\n",
    "\n",
    "        # iterate over all triplets\n",
    "        n = h.n_haplotypes\n",
    "        seen = set([hash(h[:, i].tobytes()) for i in range(h.n_haplotypes)])\n",
    "        new_haps = list()\n",
    "        for i in range(n):\n",
    "            for j in range(i + 1, n):\n",
    "                if all_edges[i, j]:\n",
    "                    for k in range(n):\n",
    "                        if all_edges[i, k] or all_edges[j, k]:\n",
    "                            if debug: print(iteration, i, j, k, 'computing median vector')\n",
    "                            uvw = h[:, [i, j, k]]\n",
    "                            ac = uvw.count_alleles(max_allele=1)\n",
    "                            # majority consensus haplotype\n",
    "                            x = np.argmax(ac, axis=1).astype('i1')\n",
    "                            x_hash = hash(x.tobytes())\n",
    "                            if debug: print(iteration, i, j, k, 'median vector', x)\n",
    "                            # test if x already in haps\n",
    "                            if x_hash in seen:\n",
    "                                if debug: print(iteration, i, j, k, 'median vector already present')\n",
    "                                pass\n",
    "                            else:\n",
    "                                if debug: print(iteration, i, j, k, 'adding median vector')\n",
    "                                new_haps.append(x.tolist())\n",
    "                                seen.add(x_hash)\n",
    "        n_medians_added = len(new_haps)\n",
    "        if debug: print(new_haps)\n",
    "        if n_medians_added:\n",
    "            h = h.concatenate(allel.HaplotypeArray(np.array(new_haps, dtype='i1').T), axis=1)\n",
    "                        \n",
    "        iteration += 1\n",
    "                \n",
    "    # final pass\n",
    "    h, edges, alt_edges = _remove_obsolete(h, orig_n_haplotypes, max_dist=max_dist, debug=debug)\n",
    "    return h, edges, alt_edges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "h = allel.HaplotypeArray([[0, 0, 1, 1, 0],\n",
    "                          [0, 0, 1, 0, 1],\n",
    "                          [0, 0, 0, 1, 1],\n",
    "                          [0, 0, 0, 1, 0],\n",
    "                          [0, 0, 0, 0, 1]])\n",
    "h = h[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# h = np.array([[0, 0, 0, 1, 1],\n",
    "#               [0, 0, 1, 0, 1],\n",
    "#               [0, 0, 1, 0, 0],\n",
    "#               [0, 0, 0, 1, 1],\n",
    "#              ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"235pt\" height=\"237pt\"\n",
       " viewBox=\"0.00 0.00 235.34 237.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(119.718 119.912)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-119.718,117.089 -119.718,-119.912 115.624,-119.912 115.624,117.089 -119.718,117.089\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-69.8428\" cy=\"-74.5201\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-69.8428\" y=\"-70.8201\" font-family=\"monospace\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0 -->\n",
       "<g id=\"node4\" class=\"node\"><title>anon_0_1_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"8.79492\" cy=\"-101.412\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_1_0 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;anon_0_1_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-47.9378,-82.011C-34.304,-86.6733 -17.1697,-92.5328 -5.31782,-96.5858\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_0_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-101.218\" cy=\"2.37349\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_2_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;anon_0_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-78.5824,-53.101C-84.022,-39.7696 -90.8582,-23.0153 -95.5869,-11.4262\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"74.7545\" cy=\"-66.6915\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"74.7545\" y=\"-62.9915\" font-family=\"monospace\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- anon_1_2_0 -->\n",
       "<g id=\"node7\" class=\"node\"><title>anon_1_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"97.1236\" cy=\"4.42215\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_2_0 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>1&#45;&#45;anon_1_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M79.6083,-51.2609C83.5145,-38.8426 88.9569,-21.5405 92.7342,-9.53217\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-3.3974\" cy=\"97.0892\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-3.3974\" y=\"100.789\" font-family=\"monospace\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0&#45;&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>anon_0_1_0&#45;&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M21.8264,-94.5523C32.9395,-88.7025 48.8932,-80.3046 60.3762,-74.2601\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1 -->\n",
       "<g id=\"node6\" class=\"node\"><title>anon_0_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-72.4053\" cy=\"68.7687\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0&#45;&#45;anon_0_2_1 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_0_2_0&#45;&#45;anon_0_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-95.387,15.8096C-90.3882,27.3288 -83.2235,43.8392 -78.2272,55.3527\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>anon_0_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-58.7716,74.3639C-47.1449,79.1355 -30.4539,85.9853 -18.4402,90.9157\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1 -->\n",
       "<g id=\"node8\" class=\"node\"><title>anon_1_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"66.19\" cy=\"69.9699\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_0&#45;&#45;anon_1_2_1 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>anon_1_2_0&#45;&#45;anon_1_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M90.8637,17.6867C85.4969,29.059 77.8047,45.3585 72.4406,56.7251\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>anon_1_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M52.4418,75.3278C40.7175,79.897 23.8863,86.4563 11.7718,91.1775\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7f8105cd8e80>"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h[:, :-1], network_method='msn', debug=False, show_node_labels=True, node_size_factor=.2, anon_width=.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"173pt\" height=\"222pt\"\n",
       " viewBox=\"0.00 0.00 173.13 221.63\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(98.8193 90.2375)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-98.8193,131.388 -98.8193,-90.2375 74.3087,-90.2375 74.3087,131.388 -98.8193,131.388\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-71.8193\" cy=\"-60.5074\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-71.8193\" y=\"-56.8074\" font-family=\"monospace\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-0.973596\" cy=\"-24.5749\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"-0.973596\" y=\"-20.8749\" font-family=\"monospace\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;3 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-50.9465,-49.9208C-39.0282,-43.876 -24.4502,-36.4821 -14.1227,-31.2441\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"54.3087\" cy=\"-70.2375\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"54.3087\" y=\"-66.5375\" font-family=\"monospace\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;3 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M41.7643,-59.8759C32.3347,-52.0872 19.5239,-41.5056 10.393,-33.9636\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"11.9479\" cy=\"111.388\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"11.9479\" y=\"115.088\" font-family=\"monospace\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- anon_2_3_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_2_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"6.5363\" cy=\"43.9318\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 2&#45;&#45;anon_2_3_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>2&#45;&#45;anon_2_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M10.6655,95.4018C9.76529,84.1812 8.57132,69.2984 7.69889,58.4235\"/>\n",
       "</g>\n",
       "<!-- anon_2_3_0&#45;&#45;3 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_2_3_0&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M4.94364,29.4033C3.66919,17.7774 1.89061,1.55293 0.616764,-10.0674\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7f810426e550>"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h[:, :-1], network_method='mjn', debug=False, show_node_labels=True, node_size_factor=.2, anon_width=.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"271pt\" height=\"377pt\"\n",
       " viewBox=\"0.00 0.00 270.62 376.80\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(137.737 188.161)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-137.737,188.635 -137.737,-188.161 132.879,-188.161 132.879,188.635 -137.737,188.635\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-79.3028\" cy=\"-15.0944\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-79.3028\" y=\"-11.3944\" font-family=\"monospace\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_0_1_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"3.40853\" cy=\"0.778701\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_1_0 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;anon_0_1_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-56.7016,-10.757C-42.1611,-7.96653 -23.6962,-4.42296 -11.0933,-2.00434\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0 -->\n",
       "<g id=\"node6\" class=\"node\"><title>anon_0_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-119.237\" cy=\"59.7449\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_2_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;anon_0_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-90.2149,5.35578C-97.2353,18.5124 -106.15,35.2198 -112.235,46.6232\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_0 -->\n",
       "<g id=\"node8\" class=\"node\"><title>anon_0_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-88.7052\" cy=\"-99.391\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_3_0 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>0&#45;&#45;anon_0_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-81.8721,-38.1288C-83.525,-52.948 -85.624,-71.7668 -87.0567,-84.6112\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"77.8915\" cy=\"14.8618\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"77.8915\" y=\"18.5618\" font-family=\"monospace\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- anon_1_2_0 -->\n",
       "<g id=\"node10\" class=\"node\"><title>anon_1_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"87.1689\" cy=\"90.8695\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_2_0 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>1&#45;&#45;anon_1_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M79.859,30.981C81.4972,44.4027 83.8098,63.3487 85.3925,76.3156\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_0 -->\n",
       "<g id=\"node12\" class=\"node\"><title>anon_1_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"114.379\" cy=\"-52.4368\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_3_0 -->\n",
       "<g id=\"edge12\" class=\"edge\"><title>1&#45;&#45;anon_1_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M85.6295,0.589481C92.0725,-11.2943 101.168,-28.0694 107.392,-39.5505\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-30.7042\" cy=\"168.635\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-30.7042\" y=\"172.335\" font-family=\"monospace\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"33.5611\" cy=\"-168.161\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"33.5611\" y=\"-164.461\" font-family=\"monospace\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0&#45;&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>anon_0_1_0&#45;&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M17.7696,3.49406C30.4515,5.89194 48.913,9.3826 62.0272,11.8622\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1 -->\n",
       "<g id=\"node7\" class=\"node\"><title>anon_0_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-97.9819\" cy=\"130.973\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0&#45;&#45;anon_0_2_1 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_0_2_0&#45;&#45;anon_0_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-115.037,73.8173C-111.311,86.3046 -105.896,104.452 -102.172,116.93\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>anon_0_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-85.3275,138.057C-73.7561,144.535 -56.6689,154.1 -44.6953,160.803\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_1 -->\n",
       "<g id=\"node9\" class=\"node\"><title>anon_0_3_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-42.8478\" cy=\"-157.875\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_0&#45;&#45;anon_0_3_1 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>anon_0_3_0&#45;&#45;anon_0_3_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-79.6453,-110.946C-71.6058,-121.199 -59.9227,-136.099 -51.8889,-146.345\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_1&#45;&#45;3 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>anon_0_3_1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-28.4759,-159.81C-15.3339,-161.579 4.07239,-164.192 17.6711,-166.022\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1 -->\n",
       "<g id=\"node11\" class=\"node\"><title>anon_1_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"44.7309\" cy=\"152.207\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_0&#45;&#45;anon_1_2_1 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>anon_1_2_0&#45;&#45;anon_1_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M78.7845,102.988C71.3446,113.741 60.5326,129.368 53.0979,140.114\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge11\" class=\"edge\"><title>anon_1_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M30.5421,155.297C17.5677,158.122 -1.59135,162.295 -15.0168,165.219\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_1 -->\n",
       "<g id=\"node13\" class=\"node\"><title>anon_1_3_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"97.6391\" cy=\"-125.111\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_0&#45;&#45;anon_1_3_1 -->\n",
       "<g id=\"edge13\" class=\"edge\"><title>anon_1_3_0&#45;&#45;anon_1_3_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M111.072,-66.7948C108.137,-79.5356 103.872,-98.0509 100.939,-110.783\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_1&#45;&#45;3 -->\n",
       "<g id=\"edge14\" class=\"edge\"><title>anon_1_3_1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M85.5865,-133.208C74.5654,-140.613 58.2908,-151.547 46.8867,-159.209\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7f8105cd8d68>"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h, network_method='msn', debug=False, show_node_labels=True, node_size_factor=.2, anon_width=.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"187pt\" height=\"291pt\"\n",
       " viewBox=\"0.00 0.00 186.99 290.70\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(86.6322 123.883)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-86.6322,166.814 -86.6322,-123.883 100.358,-123.883 100.358,166.814 -86.6322,166.814\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-59.6322\" cy=\"-31.1016\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-59.6322\" y=\"-27.4016\" font-family=\"monospace\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 4 -->\n",
       "<g id=\"node5\" class=\"node\"><title>4</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"19.1023\" cy=\"16.7785\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"19.1023\" y=\"20.4785\" font-family=\"monospace\" font-size=\"14.00\">4</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;4 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-39.7642,-19.0195C-25.1864,-10.1544 -5.83611,1.61294 6.62659,9.19179\"/>\n",
       "</g>\n",
       "<!-- 5 -->\n",
       "<g id=\"node6\" class=\"node\"><title>5</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"21.6507\" cy=\"-84.9531\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"21.6507\" y=\"-81.2531\" font-family=\"monospace\" font-size=\"14.00\">5</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;5 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>0&#45;&#45;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-40.3693,-43.8637C-24.9709,-54.0654 -3.84866,-68.0593 9.34425,-76.7998\"/>\n",
       "</g>\n",
       "<!-- 6 -->\n",
       "<g id=\"node7\" class=\"node\"><title>6</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-56.6195\" cy=\"33.8955\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"-56.6195\" y=\"37.5955\" font-family=\"monospace\" font-size=\"14.00\">6</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;6 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;6</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-58.5615,-8.00307C-58.1421,1.04605 -57.6733,11.159 -57.3015,19.1822\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"80.3577\" cy=\"-37.455\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"80.3577\" y=\"-33.755\" font-family=\"monospace\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;4 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>1&#45;&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M68.2556,-26.7402C57.2377,-16.9853 41.0966,-2.69445 30.3531,6.81747\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;5 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>1&#45;&#45;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M67.9074,-47.5281C57.6754,-55.8065 43.2866,-67.4481 33.2844,-75.5406\"/>\n",
       "</g>\n",
       "<!-- 7 -->\n",
       "<g id=\"node8\" class=\"node\"><title>7</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"56.0188\" cy=\"19.1092\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"56.0188\" y=\"22.8092\" font-family=\"monospace\" font-size=\"14.00\">7</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;7 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>1&#45;&#45;7</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73.9629,-22.5934C70.1883,-13.8211 65.4488,-2.80638 61.7887,5.69977\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"13.1244\" cy=\"146.814\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"13.1244\" y=\"150.514\" font-family=\"monospace\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 8 -->\n",
       "<g id=\"node9\" class=\"node\"><title>8</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"6.64998\" cy=\"76.923\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"6.64998\" y=\"80.623\" font-family=\"monospace\" font-size=\"14.00\">8</text>\n",
       "</g>\n",
       "<!-- 2&#45;&#45;8 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>2&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M11.6227,130.604C10.532,118.829 9.06793,103.025 8.01117,91.6171\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-66.2228\" cy=\"-103.883\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-66.2228\" y=\"-100.183\" font-family=\"monospace\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 9 -->\n",
       "<g id=\"node10\" class=\"node\"><title>9</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-14.4294\" cy=\"-36.1274\" rx=\"14.5\" ry=\"14.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"-14.4294\" y=\"-32.4274\" font-family=\"monospace\" font-size=\"14.00\">9</text>\n",
       "</g>\n",
       "<!-- 3&#45;&#45;9 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>3&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-56.4808,-91.139C-46.9866,-78.7186 -32.6689,-59.9883 -23.4345,-47.9078\"/>\n",
       "</g>\n",
       "<!-- 4&#45;&#45;8 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>4&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M16.1513,31.032C14.195,40.4805 11.6358,52.8417 9.66633,62.3541\"/>\n",
       "</g>\n",
       "<!-- 5&#45;&#45;9 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>5&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M12.9167,-73.1337C7.24403,-65.4571 -0.119247,-55.4927 -5.77587,-47.8379\"/>\n",
       "</g>\n",
       "<!-- 6&#45;&#45;8 -->\n",
       "<g id=\"edge11\" class=\"edge\"><title>6&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M-44.4205,42.1917C-33.2169,49.8109 -16.7033,61.0412 -5.51155,68.6524\"/>\n",
       "</g>\n",
       "<!-- 6&#45;&#45;9 -->\n",
       "<g id=\"edge12\" class=\"edge\"><title>6&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M-49.0767,21.3768C-41.4922,8.78866 -29.8542,-10.5269 -22.1923,-23.2433\"/>\n",
       "</g>\n",
       "<!-- 7&#45;&#45;8 -->\n",
       "<g id=\"edge13\" class=\"edge\"><title>7&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M46.5,30.2563C37.7579,40.4938 24.8724,55.5835 16.1396,65.8102\"/>\n",
       "</g>\n",
       "<!-- 7&#45;&#45;9 -->\n",
       "<g id=\"edge14\" class=\"edge\"><title>7&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M44.3857,9.98797C31.2546,-0.307759 10.0285,-16.9506 -3.01282,-27.1759\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7f8105c74358>"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h, network_method='mjn', debug=False, show_node_labels=True, node_size_factor=.2, anon_width=.4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Matplotlib integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.image as mpimg\n",
    "import io\n",
    "\n",
    "\n",
    "def plot_gv(graph, ax, size=None, desired_size=False, ratio=None, dpi=None, interpolation='bilinear', aspect='equal'):\n",
    "    fig = ax.figure\n",
    "    if size is None:\n",
    "        # try to match size to ax\n",
    "        fw, fh = fig.get_figwidth(), fig.get_figheight()\n",
    "        bbox = ax.get_position()\n",
    "        w = fw * bbox.width\n",
    "        h = fh * bbox.height\n",
    "        size = w, h\n",
    "    if dpi is None:\n",
    "        # match dpi to fig\n",
    "        dpi = fig.dpi\n",
    "\n",
    "    # set size and resolution \n",
    "    size = '%s,%s' % (w, h)\n",
    "    if desired_size:\n",
    "        size += '!'\n",
    "    kwargs = dict(\n",
    "        size=size,\n",
    "        dpi=str(dpi),\n",
    "    )\n",
    "    if ratio:\n",
    "        kwargs['ratio'] = str(ratio)\n",
    "    else:\n",
    "        kwargs['ratio'] = ''\n",
    "    graph.attr('graph', **kwargs)\n",
    "\n",
    "    # render the graph as png\n",
    "    dat = graph.pipe(format='png')\n",
    "    \n",
    "    # read the png data into an image array\n",
    "    img = mpimg.imread(io.BytesIO(dat))\n",
    "    \n",
    "    # plot the image\n",
    "    ax.imshow(img, interpolation=interpolation, aspect=aspect)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
