{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       ".container {\n",
       "    width: 100%;\n",
       "}\n",
       "div#notebook {\n",
       "    padding-top: 0;\n",
       "}\n",
       "#header-container {\n",
       "    display: none;\n",
       "}\n",
       "#header-bar {\n",
       "    display: none;\n",
       "}\n",
       "#maintoolbar {\n",
       "    display: none;\n",
       "}\n",
       "#site {\n",
       "    height: auto !important;\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%run setup.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function to support searching a scipy tree.\n",
    "\n",
    "def get_descendant(node, desc_id):\n",
    "    \"\"\"Search the descendants of the given node.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    node : scipy.cluster.hierarchy.ClusterNode\n",
    "        The ancestor node to search from.\n",
    "    desc_id : int\n",
    "        The ID of the node to search for.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    desc : scipy.cluster.hierarchy.ClusterNode\n",
    "        If a node with the given ID is not found, returns None.\n",
    "    \n",
    "    \"\"\"\n",
    "    if node.id == desc_id:\n",
    "        return node\n",
    "    if node.is_leaf():\n",
    "        return None\n",
    "    if node.left.id == desc_id:\n",
    "        return node.left\n",
    "    if node.right.id == desc_id:\n",
    "        return node.right\n",
    "    # search left\n",
    "    l = get_descendant(node.left, desc_id)\n",
    "    if l is not None:\n",
    "        return l\n",
    "    # search right\n",
    "    r = get_descendant(node.right, desc_id)\n",
    "    return r\n",
    "\n",
    "\n",
    "# monkey-patch as a method\n",
    "scipy.cluster.hierarchy.ClusterNode.get_descendant = get_descendant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fig_haplotypes_clustered(h,\n",
    "                             distance_metric='hamming',\n",
    "                             linkage_method='single',\n",
    "                             truncate_distance=0,\n",
    "                             orientation='top',\n",
    "                             subplot_ratios=(4, 2),\n",
    "                             subplot_pad=0,\n",
    "                             despine_offset=5,\n",
    "                             count_sort=True,\n",
    "                             dend_linecolor='k',\n",
    "                             cut_height=2,\n",
    "                             highlight_clusters=True,\n",
    "                             highlight_colors=None,\n",
    "                             highlight_dend=True,\n",
    "                             highlight_freq=True,\n",
    "                             highlight_alpha=0.3,\n",
    "                             label_clusters=True,\n",
    "                             dpi=None,\n",
    "                             fig=None,\n",
    "                             ):\n",
    "    \"\"\"Construct a plot of hierarchical clustering of haplotypes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    TODO\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    TODO\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # check inputs\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    \n",
    "    # compute distance matrix\n",
    "    dist = scipy.spatial.distance.pdist(h.T, metric=distance_metric)\n",
    "    if distance_metric in {'hamming', 'jaccard'}:\n",
    "        # convert distance to number of SNPs, easier to interpret\n",
    "        dist *= h.n_variants\n",
    "    \n",
    "    # compute hierarchical clustering\n",
    "    Z = scipy.cluster.hierarchy.linkage(dist, method=linkage_method)\n",
    "\n",
    "    # Z is a linkage matrix. From the scipy docs...\n",
    "    # A 4 by (n-1) matrix Z is returned. At the i-th iteration, clusters with \n",
    "    # indices Z[i, 0] and Z[i, 1] are combined to form cluster n + i. A cluster \n",
    "    # with an index less than n corresponds to one of the original observations. \n",
    "    # The distance between clusters Z[i, 0] and Z[i, 1] is given by Z[i, 2]. The \n",
    "    # fourth value Z[i, 3] represents the number of original observations in the \n",
    "    # newly formed cluster.\n",
    "    \n",
    "    # find level to truncate dendrogram\n",
    "    lastp = h.n_haplotypes - bisect.bisect_right(Z[:, 2], truncate_distance)\n",
    "    \n",
    "    # convenience variables\n",
    "    horizontal = orientation in ['left', 'right']\n",
    "    vertical = not horizontal\n",
    "    inverted = orientation in ['bottom', 'right']\n",
    "    \n",
    "    # setup figure\n",
    "    if fig is None:\n",
    "        figsize = plt.rcParams['figure.figsize']\n",
    "        if horizontal:\n",
    "            figsize = figsize[::-1]\n",
    "        fig = plt.figure(figsize=figsize, dpi=dpi)\n",
    "        \n",
    "    # setup gridspec and axes\n",
    "    if inverted:\n",
    "        subplot_ratios = subplot_ratios[::-1]\n",
    "    if horizontal:\n",
    "        gs = plt.GridSpec(nrows=1, ncols=2, width_ratios=subplot_ratios)\n",
    "    else:\n",
    "        gs = plt.GridSpec(nrows=2, ncols=1, height_ratios=subplot_ratios)\n",
    "    if inverted:\n",
    "        ax_dend = fig.add_subplot(gs[1])\n",
    "        ax_freq = fig.add_subplot(gs[0])\n",
    "    else:\n",
    "        ax_dend = fig.add_subplot(gs[0])\n",
    "        ax_freq = fig.add_subplot(gs[1])\n",
    "    if horizontal:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "    else:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "\n",
    "    # make a dendrogram\n",
    "    kwargs_dend = dict(\n",
    "        truncate_mode='lastp', \n",
    "        p=lastp,\n",
    "        show_leaf_counts=False, \n",
    "        count_sort=count_sort, \n",
    "        no_labels=True, \n",
    "        color_threshold=0, \n",
    "        above_threshold_color=dend_linecolor, \n",
    "        orientation=orientation\n",
    "    )    \n",
    "    dend = scipy.cluster.hierarchy.dendrogram(Z, ax=ax_dend, **kwargs_dend)\n",
    "    leaves = dend['leaves']\n",
    "    ax_dend_label = 'Distance'\n",
    "    if horizontal:\n",
    "        ax_dend.set_xlabel(ax_dend_label)\n",
    "        ax_dend.set_yticks([])\n",
    "    else:\n",
    "        ax_dend.set_ylabel(ax_dend_label)\n",
    "        ax_dend.set_xticks([])\n",
    "        \n",
    "    # construct a tree and compute observation counts for the dendrogram leaves\n",
    "    tree = scipy.cluster.hierarchy.to_tree(Z)\n",
    "    s = np.arange(len(leaves))\n",
    "    t = np.array([\n",
    "        1 if l < h.n_haplotypes\n",
    "        else tree.get_descendant(l).get_count()\n",
    "        for l in leaves\n",
    "    ])\n",
    "\n",
    "    # plot frequencies bar\n",
    "    ax_freq_label = 'Frequency'\n",
    "    if horizontal:\n",
    "        ax_freq.barh(s, t, height=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_ylim(0, len(leaves))\n",
    "        ax_freq.set_yticks([])\n",
    "        ax_freq.set_xlabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='x', lw=.5)\n",
    "        if orientation == 'right':\n",
    "            ax_freq.invert_xaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_xticks(ax_freq.get_xticks()[1:])\n",
    "    else:\n",
    "        ax_freq.bar(s, t, width=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_xlim(0, len(leaves))\n",
    "        ax_freq.set_xticks([])\n",
    "        ax_freq.set_ylabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='y', lw=.5)\n",
    "        if orientation == 'top':\n",
    "            ax_freq.invert_yaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_yticks(ax_freq.get_yticks()[1:])\n",
    "\n",
    "    # cut the tree\n",
    "    cut = scipy.cluster.hierarchy.cut_tree(Z, height=cut_height)[:, 0]\n",
    "    cluster_sizes = np.bincount(cut)\n",
    "    clusters = [np.nonzero(cut == i)[0] for i in range(cut.max() + 1)]\n",
    "    \n",
    "    # now the fiddly bit - we need to figure out where the clusters have\n",
    "    # ended up in the dendrogram we plotted earlier...\n",
    "    \n",
    "    # N.B., the dendrogram was truncated, so each leaf in the dendrogram\n",
    "    # may correspond to more than one original observation (i.e., haplotype).\n",
    "    # Let's build a list storing the observations for each leaf:\n",
    "    leaf_obs = [tree.get_descendant(ix).pre_order() for ix in leaves]\n",
    "    \n",
    "    # Now let's figure out for each leaf in the dendrogram, which of the clusters\n",
    "    # obtained by cutting the tree earlier does it fall into?\n",
    "    leaf_clusters = np.array([cut[l[0]] for l in leaf_obs])\n",
    "    \n",
    "    # Now let's build a data structure that reorders the clusters so they\n",
    "    # occur in the same order as in the dendrogram, and also record the indices\n",
    "    # of the start and stop leaf for each cluster:\n",
    "    cluster_spans = list()\n",
    "    c_prv = leaf_clusters[0]\n",
    "    i_start = 0\n",
    "    for i, c in enumerate(leaf_clusters[1:], 1):\n",
    "        if c != c_prv:\n",
    "            cluster_spans.append((i_start, i, clusters[c_prv]))\n",
    "            i_start = i\n",
    "        c_prv = c\n",
    "    # don't forget the last one\n",
    "    cluster_spans.append((i_start, i+1, clusters[c]))\n",
    "    \n",
    "    # OK, now figure out which clusters we want to highlight...\n",
    "    if isinstance(highlight_clusters, (list, tuple)):\n",
    "        # user has manually specified which clusters to highlight\n",
    "        pass\n",
    "    else:\n",
    "        # assume highlight_clusters is the minimum cluster size to highlight\n",
    "        min_cluster_size = int(highlight_clusters)\n",
    "        highlight_clusters = [i for i, cs in enumerate(cluster_spans) \n",
    "                              if len(cs[2]) >= min_cluster_size]\n",
    "        \n",
    "    # setup colors for highlighting clusters\n",
    "    if highlight_colors is None:\n",
    "        highlight_colors = sns.color_palette('hls', n_colors=len(highlight_clusters))\n",
    "        \n",
    "    # do the highlighting\n",
    "    for color, cix in zip(highlight_colors, highlight_clusters):\n",
    "        start, stop, _ = cluster_spans[cix]\n",
    "        if horizontal:\n",
    "            freq_spanf = ax_freq.axhspan\n",
    "            dend_patch_xy = (0, start * 10)\n",
    "            dend_patch_width = cut_height\n",
    "            dend_patch_height = (stop - start) * 10\n",
    "        else:\n",
    "            freq_spanf = ax_freq.axvspan\n",
    "            dend_patch_xy = (start * 10, 0)\n",
    "            dend_patch_width = (stop - start) * 10\n",
    "            dend_patch_height = cut_height\n",
    "        if highlight_freq:\n",
    "            freq_spanf(start, stop, color=color, alpha=highlight_alpha, zorder=-20)\n",
    "        if highlight_dend:\n",
    "            ax_dend.add_patch(plt.Rectangle(xy=dend_patch_xy, \n",
    "                                            width=dend_patch_width, \n",
    "                                            height=dend_patch_height, \n",
    "                                            color=color, alpha=highlight_alpha, \n",
    "                                            zorder=-20))\n",
    "\n",
    "    # for debugging, label the clusters\n",
    "    if label_clusters:\n",
    "        for i, (start, stop, clst) in enumerate(cluster_spans):\n",
    "            if horizontal:\n",
    "                x = max(ax_freq.get_xlim())\n",
    "                y = (start + stop) / 2\n",
    "                ha = orientation\n",
    "                va = 'center'\n",
    "            else:\n",
    "                x = (start + stop) / 2\n",
    "                y = max(ax_freq.get_ylim())\n",
    "                ha = 'center'\n",
    "                va = orientation\n",
    "            # treat label_clusters as minimum cluster size to label\n",
    "            if len(clst) >= int(label_clusters):\n",
    "                ax_freq.text(x, y, str(i), \n",
    "                             va=va, ha=ha, fontsize=6)\n",
    "    \n",
    "    # tidy up plot\n",
    "    if horizontal:\n",
    "        gs.tight_layout(fig, w_pad=subplot_pad)\n",
    "    else:\n",
    "        gs.tight_layout(fig, h_pad=subplot_pad)\n",
    "    \n",
    "    # return some useful stuff\n",
    "    return fig, ax_dend, ax_freq, cluster_spans, leaf_obs\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
