{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       ".container {\n",
       "    width: 100%;\n",
       "}\n",
       "div#notebook {\n",
       "    padding-top: 0;\n",
       "}\n",
       "#header-container {\n",
       "    display: none;\n",
       "}\n",
       "#header-bar {\n",
       "    display: none;\n",
       "}\n",
       "#maintoolbar {\n",
       "    display: none;\n",
       "}\n",
       "#site {\n",
       "    height: auto !important;\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%run setup.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hierarchical clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function to support searching a scipy tree.\n",
    "\n",
    "def get_descendant(node, desc_id):\n",
    "    \"\"\"Search the descendants of the given node.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    node : scipy.cluster.hierarchy.ClusterNode\n",
    "        The ancestor node to search from.\n",
    "    desc_id : int\n",
    "        The ID of the node to search for.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    desc : scipy.cluster.hierarchy.ClusterNode\n",
    "        If a node with the given ID is not found, returns None.\n",
    "    \n",
    "    \"\"\"\n",
    "    if node.id == desc_id:\n",
    "        return node\n",
    "    if node.is_leaf():\n",
    "        return None\n",
    "    if node.left.id == desc_id:\n",
    "        return node.left\n",
    "    if node.right.id == desc_id:\n",
    "        return node.right\n",
    "    # search left\n",
    "    l = get_descendant(node.left, desc_id)\n",
    "    if l is not None:\n",
    "        return l\n",
    "    # search right\n",
    "    r = get_descendant(node.right, desc_id)\n",
    "    return r\n",
    "\n",
    "\n",
    "# monkey-patch as a method\n",
    "scipy.cluster.hierarchy.ClusterNode.get_descendant = get_descendant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fig_haplotypes_clustered(h,\n",
    "                             distance_metric='hamming',\n",
    "                             linkage_method='single',\n",
    "                             truncate_distance=0,\n",
    "                             orientation='top',\n",
    "                             subplot_ratios=(4, 2),\n",
    "                             subplot_pad=0,\n",
    "                             despine_offset=5,\n",
    "                             count_sort=True,\n",
    "                             dend_linecolor='k',\n",
    "                             cut_height=2,\n",
    "                             highlight_clusters=True,\n",
    "                             highlight_colors=None,\n",
    "                             highlight_dend=True,\n",
    "                             highlight_freq=True,\n",
    "                             highlight_alpha=0.3,\n",
    "                             label_clusters=True,\n",
    "                             dpi=None,\n",
    "                             fig=None,\n",
    "                             ):\n",
    "    \"\"\"Construct a plot of hierarchical clustering of haplotypes.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    TODO\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    TODO\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # check inputs\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    \n",
    "    # compute distance matrix\n",
    "    dist = scipy.spatial.distance.pdist(h.T, metric=distance_metric)\n",
    "    if distance_metric in {'hamming', 'jaccard'}:\n",
    "        # convert distance to number of SNPs, easier to interpret\n",
    "        dist *= h.n_variants\n",
    "    \n",
    "    # compute hierarchical clustering\n",
    "    Z = scipy.cluster.hierarchy.linkage(dist, method=linkage_method)\n",
    "\n",
    "    # Z is a linkage matrix. From the scipy docs...\n",
    "    # A 4 by (n-1) matrix Z is returned. At the i-th iteration, clusters with \n",
    "    # indices Z[i, 0] and Z[i, 1] are combined to form cluster n + i. A cluster \n",
    "    # with an index less than n corresponds to one of the original observations. \n",
    "    # The distance between clusters Z[i, 0] and Z[i, 1] is given by Z[i, 2]. The \n",
    "    # fourth value Z[i, 3] represents the number of original observations in the \n",
    "    # newly formed cluster.\n",
    "    \n",
    "    # find level to truncate dendrogram\n",
    "    lastp = h.n_haplotypes - bisect.bisect_right(Z[:, 2], truncate_distance)\n",
    "    \n",
    "    # convenience variables\n",
    "    horizontal = orientation in ['left', 'right']\n",
    "    vertical = not horizontal\n",
    "    inverted = orientation in ['bottom', 'right']\n",
    "    \n",
    "    # setup figure\n",
    "    if fig is None:\n",
    "        figsize = plt.rcParams['figure.figsize']\n",
    "        if horizontal:\n",
    "            figsize = figsize[::-1]\n",
    "        fig = plt.figure(figsize=figsize, dpi=dpi)\n",
    "        \n",
    "    # setup gridspec and axes\n",
    "    if inverted:\n",
    "        subplot_ratios = subplot_ratios[::-1]\n",
    "    if horizontal:\n",
    "        gs = plt.GridSpec(nrows=1, ncols=2, width_ratios=subplot_ratios)\n",
    "    else:\n",
    "        gs = plt.GridSpec(nrows=2, ncols=1, height_ratios=subplot_ratios)\n",
    "    if inverted:\n",
    "        ax_dend = fig.add_subplot(gs[1])\n",
    "        ax_freq = fig.add_subplot(gs[0])\n",
    "    else:\n",
    "        ax_dend = fig.add_subplot(gs[0])\n",
    "        ax_freq = fig.add_subplot(gs[1])\n",
    "    if horizontal:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=True, top=True, right=True, bottom=False)\n",
    "    else:\n",
    "        sns.despine(ax=ax_dend, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "        sns.despine(ax=ax_freq, offset=despine_offset, \n",
    "                    left=False, top=True, right=True, bottom=True)\n",
    "\n",
    "    # make a dendrogram\n",
    "    kwargs_dend = dict(\n",
    "        truncate_mode='lastp', \n",
    "        p=lastp,\n",
    "        show_leaf_counts=False, \n",
    "        count_sort=count_sort, \n",
    "        no_labels=True, \n",
    "        color_threshold=0, \n",
    "        above_threshold_color=dend_linecolor, \n",
    "        orientation=orientation\n",
    "    )    \n",
    "    dend = scipy.cluster.hierarchy.dendrogram(Z, ax=ax_dend, **kwargs_dend)\n",
    "    leaves = dend['leaves']\n",
    "    ax_dend_label = 'Distance'\n",
    "    if horizontal:\n",
    "        ax_dend.set_xlabel(ax_dend_label)\n",
    "        ax_dend.set_yticks([])\n",
    "    else:\n",
    "        ax_dend.set_ylabel(ax_dend_label)\n",
    "        ax_dend.set_xticks([])\n",
    "        \n",
    "    # construct a tree and compute observation counts for the dendrogram leaves\n",
    "    tree = scipy.cluster.hierarchy.to_tree(Z)\n",
    "    s = np.arange(len(leaves))\n",
    "    t = np.array([\n",
    "        1 if l < h.n_haplotypes\n",
    "        else tree.get_descendant(l).get_count()\n",
    "        for l in leaves\n",
    "    ])\n",
    "\n",
    "    # plot frequencies bar\n",
    "    ax_freq_label = 'Frequency'\n",
    "    if horizontal:\n",
    "        ax_freq.barh(s, t, height=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_ylim(0, len(leaves))\n",
    "        ax_freq.set_yticks([])\n",
    "        ax_freq.set_xlabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='x', lw=.5)\n",
    "        if orientation == 'right':\n",
    "            ax_freq.invert_xaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_xticks(ax_freq.get_xticks()[1:])\n",
    "    else:\n",
    "        ax_freq.bar(s, t, width=1, lw=0, color='k', align='edge')\n",
    "        ax_freq.set_xlim(0, len(leaves))\n",
    "        ax_freq.set_xticks([])\n",
    "        ax_freq.set_ylabel(ax_freq_label)\n",
    "        ax_freq.grid(axis='y', lw=.5)\n",
    "        if orientation == 'top':\n",
    "            ax_freq.invert_yaxis()\n",
    "        # remove 0\n",
    "        ax_freq.set_yticks(ax_freq.get_yticks()[1:])\n",
    "\n",
    "    # cut the tree\n",
    "    cut = scipy.cluster.hierarchy.cut_tree(Z, height=cut_height)[:, 0]\n",
    "    cluster_sizes = np.bincount(cut)\n",
    "    clusters = [np.nonzero(cut == i)[0] for i in range(cut.max() + 1)]\n",
    "    \n",
    "    # now the fiddly bit - we need to figure out where the clusters have\n",
    "    # ended up in the dendrogram we plotted earlier...\n",
    "    \n",
    "    # N.B., the dendrogram was truncated, so each leaf in the dendrogram\n",
    "    # may correspond to more than one original observation (i.e., haplotype).\n",
    "    # Let's build a list storing the observations for each leaf:\n",
    "    leaf_obs = [tree.get_descendant(ix).pre_order() for ix in leaves]\n",
    "    \n",
    "    # Now let's figure out for each leaf in the dendrogram, which of the clusters\n",
    "    # obtained by cutting the tree earlier does it fall into?\n",
    "    leaf_clusters = np.array([cut[l[0]] for l in leaf_obs])\n",
    "    \n",
    "    # Now let's build a data structure that reorders the clusters so they\n",
    "    # occur in the same order as in the dendrogram, and also record the indices\n",
    "    # of the start and stop leaf for each cluster:\n",
    "    cluster_spans = list()\n",
    "    c_prv = leaf_clusters[0]\n",
    "    i_start = 0\n",
    "    for i, c in enumerate(leaf_clusters[1:], 1):\n",
    "        if c != c_prv:\n",
    "            cluster_spans.append((i_start, i, clusters[c_prv]))\n",
    "            i_start = i\n",
    "        c_prv = c\n",
    "    # don't forget the last one\n",
    "    cluster_spans.append((i_start, i+1, clusters[c]))\n",
    "    \n",
    "    # OK, now figure out which clusters we want to highlight...\n",
    "    if isinstance(highlight_clusters, (list, tuple)):\n",
    "        # user has manually specified which clusters to highlight\n",
    "        pass\n",
    "    else:\n",
    "        # assume highlight_clusters is the minimum cluster size to highlight\n",
    "        min_cluster_size = int(highlight_clusters)\n",
    "        highlight_clusters = [i for i, cs in enumerate(cluster_spans) \n",
    "                              if len(cs[2]) >= min_cluster_size]\n",
    "        \n",
    "    # setup colors for highlighting clusters\n",
    "    if highlight_colors is None:\n",
    "        highlight_colors = sns.color_palette('hls', n_colors=len(highlight_clusters))\n",
    "        \n",
    "    # do the highlighting\n",
    "    for color, cix in zip(highlight_colors, highlight_clusters):\n",
    "        start, stop, _ = cluster_spans[cix]\n",
    "        if horizontal:\n",
    "            freq_spanf = ax_freq.axhspan\n",
    "            dend_patch_xy = (0, start * 10)\n",
    "            dend_patch_width = cut_height\n",
    "            dend_patch_height = (stop - start) * 10\n",
    "        else:\n",
    "            freq_spanf = ax_freq.axvspan\n",
    "            dend_patch_xy = (start * 10, 0)\n",
    "            dend_patch_width = (stop - start) * 10\n",
    "            dend_patch_height = cut_height\n",
    "        if highlight_freq:\n",
    "            freq_spanf(start, stop, color=color, alpha=highlight_alpha, zorder=-20)\n",
    "        if highlight_dend:\n",
    "            ax_dend.add_patch(plt.Rectangle(xy=dend_patch_xy, \n",
    "                                            width=dend_patch_width, \n",
    "                                            height=dend_patch_height, \n",
    "                                            color=color, alpha=highlight_alpha, \n",
    "                                            zorder=-20))\n",
    "\n",
    "    # for debugging, label the clusters\n",
    "    if label_clusters:\n",
    "        for i, (start, stop, clst) in enumerate(cluster_spans):\n",
    "            if horizontal:\n",
    "                x = max(ax_freq.get_xlim())\n",
    "                y = (start + stop) / 2\n",
    "                ha = orientation\n",
    "                va = 'center'\n",
    "            else:\n",
    "                x = (start + stop) / 2\n",
    "                y = max(ax_freq.get_ylim())\n",
    "                ha = 'center'\n",
    "                va = orientation\n",
    "            # treat label_clusters as minimum cluster size to label\n",
    "            if len(clst) >= int(label_clusters):\n",
    "                ax_freq.text(x, y, str(i), \n",
    "                             va=va, ha=ha, fontsize=6)\n",
    "    \n",
    "    # tidy up plot\n",
    "    if horizontal:\n",
    "        gs.tight_layout(fig, w_pad=subplot_pad)\n",
    "    else:\n",
    "        gs.tight_layout(fig, h_pad=subplot_pad)\n",
    "    \n",
    "    # return some useful stuff\n",
    "    return fig, ax_dend, ax_freq, cluster_spans, leaf_obs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotype networks (via minimum spanning tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _graph_edges(graph, \n",
    "                 edges, \n",
    "                 hap_counts, \n",
    "                 node_size_factor, \n",
    "                 edge_length,\n",
    "                 anon_width,\n",
    "                 intermediate_nodes,\n",
    "                 edge_attrs,\n",
    "                 anon_node_attrs):\n",
    "    \n",
    "    for i in range(edges.shape[0]):\n",
    "\n",
    "        for j in range(edges.shape[1]):\n",
    "\n",
    "            # lookup distance between nodes i and j\n",
    "            sep = edges[i, j]\n",
    "\n",
    "            if sep > 0:\n",
    "\n",
    "                # lookup number of haplotypes\n",
    "                # calculate node sizes (needed to adjust edge length)\n",
    "                if i < len(hap_counts):\n",
    "                    n_i = hap_counts[i]\n",
    "                    width_i = np.sqrt(n_i * node_size_factor)\n",
    "                else:\n",
    "                    n_i = 1\n",
    "                    width_i = anon_width\n",
    "                if j < len(hap_counts):\n",
    "                    n_j = hap_counts[j]\n",
    "                    width_j = np.sqrt(n_j * node_size_factor)\n",
    "                else:\n",
    "                    n_j = 1\n",
    "                    width_j = anon_width\n",
    "\n",
    "                if sep > 1 and intermediate_nodes:\n",
    "\n",
    "                    # tricky case, need to add some anonymous nodes to represent intermediate steps\n",
    "\n",
    "                    # add first intermediate node\n",
    "                    nid = 'anon_{}_{}_{}'.format(i, j, 0)\n",
    "                    graph.node(nid, label='', width=str(anon_width), **anon_node_attrs)\n",
    "\n",
    "                    # add edge from node i to first intermediate\n",
    "                    el = edge_length + width_i / 2 + anon_width / 2\n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge(str(i), 'anon_{}_{}_{}'.format(i, j, 0), **kwargs)\n",
    "\n",
    "                    # add further intermediate nodes as necessary\n",
    "                    for k in range(1, sep-1):\n",
    "                        nid = 'anon_{}_{}_{}'.format(i, j, k)\n",
    "                        graph.node(nid, label='', width=str(anon_width), **anon_node_attrs)\n",
    "                        el = edge_length + anon_width\n",
    "                        kwargs = {'len': str(el)}\n",
    "                        kwargs.update(edge_attrs)\n",
    "                        graph.edge('anon_{}_{}_{}'.format(i, j, k-1), 'anon_{}_{}_{}'.format(i, j, k), **kwargs)\n",
    "\n",
    "                    # add edge from final intermediate node to node j\n",
    "                    el = edge_length + anon_width / 2 + width_j / 2 \n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge('anon_{}_{}_{}'.format(i, j, sep-2), str(j), **kwargs)\n",
    "\n",
    "                else:\n",
    "\n",
    "                    # simple case, direct edge from node i to j\n",
    "\n",
    "                    # N.B., adjust edge length so we measure distance from edge of circle rather than center\n",
    "                    el = (edge_length * sep) + width_i / 2 + width_j / 2\n",
    "                    kwargs = {'len': str(el)}\n",
    "                    kwargs.update(edge_attrs)\n",
    "                    graph.edge(str(i), str(j), **kwargs)\n",
    "\n",
    "                    \n",
    "def graph_haplotype_network(h,\n",
    "                            hap_colors='grey',\n",
    "                            distance_metric='hamming',\n",
    "                            network_method='mst',\n",
    "                            comment=None,\n",
    "                            engine='neato',\n",
    "                            format='png',\n",
    "                            mode='major',\n",
    "                            overlap=True,\n",
    "                            splines=False,\n",
    "                            node_size_factor=0.01,\n",
    "                            node_attrs=None,\n",
    "                            show_node_labels=False,\n",
    "                            edge_length=0.5,\n",
    "                            edge_weight=10,\n",
    "                            edge_attrs=None,\n",
    "                            show_alternate_edges=True,\n",
    "                            alternate_edge_attrs=None,\n",
    "                            anon_width=0.07,\n",
    "                            anon_fillcolor='white',\n",
    "                            anon_node_attrs=None,\n",
    "                            intermediate_nodes=True,\n",
    "                            max_dist=5,\n",
    "                            debug=False,\n",
    "                            ):\n",
    "    \"\"\"TODO doc me\"\"\"\n",
    "    \n",
    "    # check inputs\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    \n",
    "    # find distinct haplotypes\n",
    "    h_distinct_sets = h.distinct()\n",
    "    \n",
    "    # find indices of distinct haplotypes - just need one per set\n",
    "    h_distinct_indices = [sorted(s)[0] for s in h_distinct_sets]\n",
    "    \n",
    "    # reorder by index\n",
    "    ix = np.argsort(h_distinct_indices)\n",
    "    h_distinct_indices = [h_distinct_indices[i] for i in ix]\n",
    "    h_distinct_sets = [h_distinct_sets[i] for i in ix]\n",
    "    \n",
    "    # obtain an array of distinct haplotypes\n",
    "    h_distinct = h.take(h_distinct_indices, axis=1)\n",
    "\n",
    "    # deal with colors - count how many of each color per distinct haplotype\n",
    "    color_counters = None\n",
    "    if isinstance(hap_colors, (list, tuple, np.ndarray)):\n",
    "        assert len(hap_colors) == h.n_haplotypes\n",
    "        color_counters = [\n",
    "            collections.Counter([hap_colors[i] for i in s])\n",
    "            for s in h_distinct_sets\n",
    "        ]\n",
    "\n",
    "    # count how many observations per distinct haplotype\n",
    "    hap_counts = [len(s) for s in h_distinct_sets]\n",
    "\n",
    "    # compute pairwise distance matrix\n",
    "    assert distance_metric in ['hamming', 'jaccard']\n",
    "    dist = allel.pairwise_distance(h_distinct, metric=distance_metric)\n",
    "    dist *= h.n_variants\n",
    "    dist = scipy.spatial.distance.squareform(dist).astype(int)\n",
    "\n",
    "    if network_method.lower() == 'mst':\n",
    "\n",
    "        # compute minimum spanning tree\n",
    "        edges = scipy.sparse.csgraph.minimum_spanning_tree(dist).toarray().astype(int)\n",
    "\n",
    "        # deal with maximum distance\n",
    "        if max_dist:\n",
    "            edges[edges > max_dist] = 0\n",
    "\n",
    "        # no alternate edges when using mst\n",
    "        alternate_edges = None\n",
    "        \n",
    "    elif network_method.lower() == 'msn':\n",
    "        \n",
    "        # compute network\n",
    "        edges, alternate_edges = minimum_spanning_network(dist, max_dist=max_dist, debug=debug)\n",
    "        edges = np.triu(edges)\n",
    "        alternate_edges = np.triu(alternate_edges)\n",
    "        \n",
    "    elif network_method.lower() == 'mjn':\n",
    "        \n",
    "        # compute network\n",
    "        _, edges, alternate_edges = median_joining_network(h_distinct, max_dist=max_dist, debug=debug)\n",
    "        edges = np.triu(edges)\n",
    "        alternate_edges = np.triu(alternate_edges)\n",
    "        \n",
    "    else:\n",
    "        raise ValueError(network_method)\n",
    "\n",
    "    # setup graph\n",
    "    graph = graphviz.Graph(comment=comment, engine=engine, format=format)\n",
    "    graph.attr('graph', overlap=str(overlap).lower(), splines=str(splines).lower(), mode=mode)\n",
    "\n",
    "    # add the main nodes\n",
    "    if node_attrs is None:\n",
    "        node_attrs = dict()\n",
    "    node_attrs.setdefault('fixedsize', 'true')\n",
    "    node_attrs.setdefault('shape', 'circle')\n",
    "    if anon_node_attrs is None:\n",
    "        anon_node_attrs = dict()\n",
    "    anon_node_attrs.setdefault('fixedsize', 'true')\n",
    "    anon_node_attrs.setdefault('shape', 'circle')\n",
    "    anon_node_attrs.setdefault('style', 'filled')\n",
    "    anon_node_attrs.setdefault('fillcolor', anon_fillcolor)\n",
    "    for i in range(edges.shape[0]):\n",
    "        kwargs = dict()\n",
    "        \n",
    "        if i < len(hap_counts):\n",
    "            n = hap_counts[i]\n",
    "\n",
    "            # calculate width from number of items - make width proportional to area\n",
    "            width = np.sqrt(n * node_size_factor)\n",
    "\n",
    "            # determine style and fill color\n",
    "            if color_counters:\n",
    "                cc = color_counters[i]\n",
    "                if len(cc) > 1:\n",
    "                    # more than one color, make a pie chart\n",
    "                    style = 'wedged'\n",
    "                    fillcolor = ':'.join(['%s;%s' % (k, v/n) for k, v in cc.items()])\n",
    "                else:\n",
    "                    # just one color, fill with solid color\n",
    "                    style = 'filled'\n",
    "                    fillcolor = list(cc.keys())[0]\n",
    "            else:\n",
    "                style = 'filled'\n",
    "                fillcolor = hap_colors\n",
    "            \n",
    "            kwargs.update(node_attrs)\n",
    "            kwargs.setdefault('style', style)\n",
    "            kwargs.setdefault('fillcolor', fillcolor)\n",
    "            kwargs.setdefault('width', str(width))\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            width = anon_width\n",
    "            fillcolor = anon_fillcolor\n",
    "            kwargs.update(anon_node_attrs)\n",
    "            kwargs.setdefault('width', str(anon_width))\n",
    "\n",
    "        # add graph node\n",
    "        if show_node_labels:\n",
    "            label = str(i)\n",
    "        else:\n",
    "            label = \"\"\n",
    "        kwargs.setdefault('label', label)\n",
    "        graph.node(str(i), **kwargs)\n",
    "    \n",
    "    # setup defaults\n",
    "    if edge_attrs is None:\n",
    "        edge_attrs = dict()\n",
    "    edge_attrs.setdefault('style', 'normal')\n",
    "    edge_attrs.setdefault('weight', str(edge_weight))\n",
    "    if alternate_edge_attrs is None:\n",
    "        alternate_edge_attrs = dict()\n",
    "    alternate_edge_attrs.setdefault('style', 'dashed')\n",
    "    alternate_edge_attrs.setdefault('weight', str(edge_weight))\n",
    "\n",
    "    # add main edges\n",
    "    _graph_edges(graph, \n",
    "                 edges, \n",
    "                 hap_counts, \n",
    "                 node_size_factor, \n",
    "                 edge_length,\n",
    "                 anon_width,\n",
    "                 intermediate_nodes,\n",
    "                 edge_attrs,\n",
    "                 anon_node_attrs)\n",
    "    \n",
    "    # add alternate edges\n",
    "    if show_alternate_edges and alternate_edges is not None:\n",
    "        _graph_edges(graph, \n",
    "                     alternate_edges, \n",
    "                     hap_counts, \n",
    "                     node_size_factor, \n",
    "                     edge_length,\n",
    "                     anon_width,\n",
    "                     intermediate_nodes,\n",
    "                     alternate_edge_attrs,\n",
    "                     anon_node_attrs)\n",
    "\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def minimum_spanning_network(dist, max_dist=None, debug=False):\n",
    "    \"\"\"TODO\"\"\"\n",
    "    \n",
    "    # TODO review implementation, see if this can be tidied up\n",
    "    \n",
    "    # keep only the lower triangle of the distance matrix, to avoid adding the same\n",
    "    # edge twice\n",
    "    dist = np.triu(dist)\n",
    "    \n",
    "    # setup the output array of links between nodes\n",
    "    edges = np.zeros_like(dist)\n",
    "    \n",
    "    # setup an array of alternate links\n",
    "    alternate_edges = np.zeros_like(dist)\n",
    "    \n",
    "    # intermediate variable - assignment of haplotypes to clusters (a.k.a. sub-networks)\n",
    "    # initially each distinct haplotype is in its own cluster\n",
    "    cluster = np.arange(dist.shape[0])\n",
    "    \n",
    "    # start with haplotypes separated by a single mutation\n",
    "    step = 1\n",
    "    \n",
    "    # iterate until all haplotypes in a single cluster, or max_dist reached\n",
    "    while len(set(cluster)) > 1 and (max_dist is None or step <= max_dist):\n",
    "        if debug: print('processing step', step, cluster)\n",
    "            \n",
    "        # keep track of which clusters have been merged at this height\n",
    "        merged = set()\n",
    "        \n",
    "        # remember what cluster assignments were at the previous height\n",
    "        prv_cluster = cluster.copy()\n",
    "        \n",
    "        # iterate over all pairs where distance equals current step size\n",
    "        for i, j in zip(*np.nonzero(dist == step)):\n",
    "            if debug: print('found potential edge', i, j)\n",
    "            \n",
    "            # current cluster assignment for each haplotype\n",
    "            a = cluster[i]\n",
    "            b = cluster[j]\n",
    "            \n",
    "            # previous cluster assignment for each haplotype\n",
    "            pa = prv_cluster[i]\n",
    "            pb = prv_cluster[j]\n",
    "            \n",
    "            if debug: print(a, b, pa, pb, merged)\n",
    "            \n",
    "            # check to see if both nodes already in the same cluster\n",
    "            if a != b:\n",
    "                \n",
    "                # nodes are in different clusters, so we can merge (i.e., connect) the clusters\n",
    "                \n",
    "                if debug: print('assign an edge')\n",
    "                edges[i, j] = dist[i, j]\n",
    "                edges[j, i] = dist[i, j]\n",
    "                \n",
    "                # merge clusters\n",
    "                c = cluster.max() + 1\n",
    "                loc_a = cluster == a\n",
    "                loc_b = cluster == b\n",
    "                cluster[loc_a] = c\n",
    "                cluster[loc_b] = c\n",
    "                merged.add(tuple(sorted([pa, pb])))\n",
    "                if debug: print('merged', cluster, merged)\n",
    "\n",
    "            elif tuple(sorted([pa, pb])) in merged or step == 1:\n",
    "                \n",
    "                # the two clusters have already been merged at this level, this is an alternate connection\n",
    "                # N.B., special case step = 1 because no previous cluster assignments\n",
    "                \n",
    "                if debug: print('assign an alternate edge')\n",
    "                alternate_edges[i, j] = dist[i, j]\n",
    "                alternate_edges[j, i] = dist[i, j]\n",
    "                \n",
    "            else:\n",
    "                \n",
    "                if debug: print('')\n",
    "                    \n",
    "        # increment step\n",
    "        step += 1\n",
    "                \n",
    "    return edges, alternate_edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _remove_obsolete(h, orig_n_haplotypes, max_dist=None, debug=False):\n",
    "    n_removed = None\n",
    "    \n",
    "    while n_removed is None or n_removed > 0:\n",
    "\n",
    "        # step 1 - compute distance\n",
    "        dist = allel.pairwise_distance(h, metric='hamming')\n",
    "        dist *= h.n_variants\n",
    "        dist = scipy.spatial.distance.squareform(dist).astype(int)\n",
    "\n",
    "        # step 2 - construct the minimum spanning network\n",
    "        edges, alt_edges = minimum_spanning_network(dist, max_dist=max_dist)\n",
    "        all_edges = edges + alt_edges\n",
    "\n",
    "        # step 3 - remove obsolete sequence types\n",
    "        loc_keep = np.ones(h.n_haplotypes, dtype=bool)\n",
    "        for i in range(orig_n_haplotypes, h.n_haplotypes):\n",
    "            n_connections = np.count_nonzero(all_edges[i])\n",
    "            if n_connections <= 2:\n",
    "                loc_keep[i] = False\n",
    "        n_removed = np.count_nonzero(~loc_keep)\n",
    "        if debug: print('discarding', n_removed, 'obsolete haplotypes')\n",
    "        h = h[:, loc_keep]\n",
    "        \n",
    "    return h, edges, alt_edges\n",
    "    \n",
    "\n",
    "def median_joining_network(h, debug=False, max_dist=None):\n",
    "    h = allel.HaplotypeArray(h)\n",
    "    orig_n_haplotypes = h.n_haplotypes\n",
    "    \n",
    "    n_medians_added = None\n",
    "    iteration = 0\n",
    "    while n_medians_added is None or n_medians_added > 0:\n",
    "        n_medians_added = 0\n",
    "\n",
    "        # steps 1-3\n",
    "        h, edges, alt_edges = _remove_obsolete(h, orig_n_haplotypes, max_dist=max_dist, debug=debug)\n",
    "        all_edges = edges + alt_edges\n",
    "\n",
    "        # step 4 - add median vectors\n",
    "\n",
    "        # iterate over all triplets\n",
    "        for i, j, k in itertools.combinations(range(h.n_haplotypes), 3): \n",
    "            if debug: print(iteration, i, j, k)\n",
    "            n_pairs_linked = 0\n",
    "            n_pairs_linked += all_edges[i, j] > 0\n",
    "            n_pairs_linked += all_edges[i, k] > 0\n",
    "            n_pairs_linked += all_edges[j, k] > 0\n",
    "            if debug: print(iteration, i, j, k, 'n_pairs_linked', n_pairs_linked)\n",
    "            if n_pairs_linked >= 2:\n",
    "                if debug: print(iteration, i, j, k, 'computing median vector')\n",
    "                uvw = h[:, [i, j, k]]\n",
    "                ac = uvw.count_alleles(max_allele=1)\n",
    "                # majority consensus haplotype\n",
    "                x = np.argmax(ac, axis=1)\n",
    "                if debug: print(iteration, i, j, k, 'median vector', x)\n",
    "                # test if x already in haps\n",
    "                if np.any(np.all(x[:, None] == np.asarray(h), axis=0)):\n",
    "                    if debug: print(iteration, i, j, k, 'median vector already present')\n",
    "                    pass\n",
    "                else:\n",
    "                    if debug: print(iteration, i, j, k, 'adding median vector')\n",
    "                    h = allel.HaplotypeArray(np.column_stack([h, x]))\n",
    "                    n_medians_added += 1\n",
    "            if debug: print(iteration, i, j, k, 'n_medians_added', n_medians_added)\n",
    "                \n",
    "        iteration += 1\n",
    "                \n",
    "    # final pass\n",
    "    h, edges, alt_edges = _remove_obsolete(h, orig_n_haplotypes, max_dist=max_dist, debug=debug)\n",
    "    return h, edges, alt_edges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "h = allel.HaplotypeArray([[0, 0, 1, 1, 0],\n",
    "                          [0, 0, 1, 0, 1],\n",
    "                          [0, 0, 0, 1, 1],\n",
    "                          [0, 0, 0, 1, 0],\n",
    "                          [0, 0, 0, 0, 1]])\n",
    "h = h[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# h = np.array([[0, 0, 0, 1, 1],\n",
    "#               [0, 0, 1, 0, 1],\n",
    "#               [0, 0, 1, 0, 0],\n",
    "#               [0, 0, 0, 1, 1],\n",
    "#              ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"169pt\" height=\"179pt\"\n",
       " viewBox=\"0.00 0.00 169.20 179.17\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(86.1071 87.1522)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-86.1071,92.0213 -86.1071,-87.1522 83.0942,-87.1522 83.0942,92.0213 -86.1071,92.0213\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-56.2796\" cy=\"-59.2245\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-56.2796\" y=\"-55.5245\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0 -->\n",
       "<g id=\"node4\" class=\"node\"><title>anon_0_1_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"10.5911\" cy=\"-80.6522\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_1_0 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;anon_0_1_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-34.0126,-66.3596C-18.2073,-71.4242 1.3598,-77.6942 8.1418,-79.8674\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_0_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-79.6071\" cy=\"6.64931\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_2_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;anon_0_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-64.0473,-37.2895C-69.5609,-21.7198 -76.3868,-2.44436 -78.7527,4.23653\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"63.0942\" cy=\"-48.5455\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"63.0942\" y=\"-44.8455\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- anon_1_2_0 -->\n",
       "<g id=\"node7\" class=\"node\"><title>anon_1_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"74.8477\" cy=\"11.7128\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_2_0 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>1&#45;&#45;anon_1_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M66.1823,-32.7133C69.0117,-18.2074 72.9638,2.05456 74.3464,9.14281\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-4.73429\" cy=\"72.0213\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-4.73429\" y=\"75.7213\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0&#45;&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>anon_0_1_0&#45;&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M12.8471,-79.2727C19.0495,-75.4798 36.736,-64.6641 49.3682,-56.9392\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1 -->\n",
       "<g id=\"node6\" class=\"node\"><title>anon_0_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-61.0419\" cy=\"47.2264\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0&#45;&#45;anon_0_2_1 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_0_2_0&#45;&#45;anon_0_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-78.389,9.31169C-74.9686,16.7874 -65.3168,37.883 -62.1,44.9138\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>anon_0_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-58.6224,48.2918C-51.9706,51.2209 -33.0025,59.5734 -19.4549,65.5391\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1 -->\n",
       "<g id=\"node8\" class=\"node\"><title>anon_1_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"53.1299\" cy=\"50.8125\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_0&#45;&#45;anon_1_2_1 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>anon_1_2_0&#45;&#45;anon_1_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M73.4227,14.2782C69.4215,21.4818 58.1306,41.8093 54.3676,48.5841\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>anon_1_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M50.6435,51.7238C43.8078,54.2293 24.3154,61.3738 10.3933,66.4766\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7fc2166b82e8>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h[:, :-1], network_method='msn', debug=False, show_node_labels=True, node_size_factor=.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"152pt\" height=\"177pt\"\n",
       " viewBox=\"0.00 0.00 152.09 177.23\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(85.6462 73.1614)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-85.6462,104.07 -85.6462,-73.1614 66.4389,-73.1614 66.4389,104.07 -85.6462,104.07\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-58.6462\" cy=\"-46.1614\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-58.6462\" y=\"-42.4614\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"0.298975\" cy=\"-14.6727\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"0.298975\" y=\"-10.9727\" font-family=\"Times,serif\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;3 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-38.03,-35.1482C-24.4847,-27.9122 -8.18795,-19.2065 -2.12627,-15.9683\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"46.4389\" cy=\"-51.5953\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"46.4389\" y=\"-47.8953\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;3 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M33.831,-41.506C22.9447,-32.7945 8.11529,-20.9276 2.53999,-16.4661\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"7.90921\" cy=\"84.0701\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"7.90921\" y=\"87.7701\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- anon_2_3_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_2_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"3.99911\" cy=\"28.3594\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 2&#45;&#45;anon_2_3_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>2&#45;&#45;anon_2_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M6.77825,67.9562C5.85802,54.8448 4.63962,37.4853 4.18262,30.974\"/>\n",
       "</g>\n",
       "<!-- anon_2_3_0&#45;&#45;3 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_2_3_0&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M3.77873,25.7964C3.11907,18.1246 1.16559,-4.59415 0.513564,-12.1771\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7fc216664ba8>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h[:, :-1], network_method='mjn', debug=False, show_node_labels=True, node_size_factor=.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"188pt\" height=\"296pt\"\n",
       " viewBox=\"0.00 0.00 188.21 296.34\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(95.9767 148.03)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-95.9767,148.311 -95.9767,-148.03 92.2366,-148.03 92.2366,148.311 -95.9767,148.311\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-67.2045\" cy=\"-6.96247\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-67.2045\" y=\"-3.26247\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0 -->\n",
       "<g id=\"node5\" class=\"node\"><title>anon_0_1_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"2.96668\" cy=\"0.415719\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_1_0 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;anon_0_1_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-44.2278,-4.54658C-27.6001,-2.79825 -6.83669,-0.615062 0.365027,0.142167\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0 -->\n",
       "<g id=\"node6\" class=\"node\"><title>anon_0_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-89.4767\" cy=\"60.8175\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_2_0 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;anon_0_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-74.4972,15.2312C-79.7748,31.2923 -86.3651,51.3482 -88.6509,58.3045\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_0 -->\n",
       "<g id=\"node8\" class=\"node\"><title>anon_0_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-74.7974\" cy=\"-77.8518\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;anon_0_3_0 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>0&#45;&#45;anon_0_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-69.6907,-30.1743C-71.4899,-46.9722 -73.7367,-67.9481 -74.5159,-75.2235\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"64.8466\" cy=\"6.75169\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"64.8466\" y=\"10.4517\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- anon_1_2_0 -->\n",
       "<g id=\"node10\" class=\"node\"><title>anon_1_2_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73.2169\" cy=\"69.2911\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_2_0 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>1&#45;&#45;anon_1_2_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M67.0021,22.8571C69.0172,37.9127 71.8605,59.1566 72.8559,66.5938\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_0 -->\n",
       "<g id=\"node12\" class=\"node\"><title>anon_1_3_0</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"85.7366\" cy=\"-52.7426\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;anon_1_3_0 -->\n",
       "<g id=\"edge12\" class=\"edge\"><title>1&#45;&#45;anon_1_3_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M70.2263,-8.56958C75.2553,-22.8921 82.3514,-43.1017 84.8356,-50.1767\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-11.7997\" cy=\"128.311\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-11.7997\" y=\"132.011\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"14.8944\" cy=\"-128.03\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"14.8944\" y=\"-124.33\" font-family=\"Times,serif\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- anon_0_1_0&#45;&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>anon_0_1_0&#45;&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M5.62558,0.687967C12.9357,1.43646 33.7809,3.57083 48.6692,5.09527\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1 -->\n",
       "<g id=\"node7\" class=\"node\"><title>anon_0_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-70.3267\" cy=\"102.38\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_0&#45;&#45;anon_0_2_1 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>anon_0_2_0&#45;&#45;anon_0_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-88.2202,63.5445C-84.6921,71.2019 -74.7362,92.81 -71.4181,100.012\"/>\n",
       "</g>\n",
       "<!-- anon_0_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>anon_0_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-67.8119,103.495C-60.8104,106.597 -40.681,115.515 -26.5685,121.767\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_1 -->\n",
       "<g id=\"node9\" class=\"node\"><title>anon_0_3_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-47.7566\" cy=\"-114.749\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_0&#45;&#45;anon_0_3_1 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>anon_0_3_0&#45;&#45;anon_0_3_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-73.0232,-80.2727C-68.0413,-87.0706 -53.9831,-106.253 -49.2977,-112.647\"/>\n",
       "</g>\n",
       "<!-- anon_0_3_1&#45;&#45;3 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>anon_0_3_1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-45.0646,-115.32C-37.5697,-116.909 -16.022,-121.476 -0.915127,-124.678\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1 -->\n",
       "<g id=\"node11\" class=\"node\"><title>anon_1_2_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"49.1767\" cy=\"108.459\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_0&#45;&#45;anon_1_2_1 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>anon_1_2_0&#45;&#45;anon_1_2_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M71.6395,71.861C67.2105,79.0772 54.7123,99.4404 50.5468,106.227\"/>\n",
       "</g>\n",
       "<!-- anon_1_2_1&#45;&#45;2 -->\n",
       "<g id=\"edge11\" class=\"edge\"><title>anon_1_2_1&#45;&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M46.5566,109.312C39.2621,111.687 18.2903,118.515 3.58721,123.301\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_1 -->\n",
       "<g id=\"node13\" class=\"node\"><title>anon_1_3_1</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"70.5239\" cy=\"-96.0904\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_0&#45;&#45;anon_1_3_1 -->\n",
       "<g id=\"edge13\" class=\"edge\"><title>anon_1_3_0&#45;&#45;anon_1_3_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M84.8305,-55.3244C82.1184,-63.0525 74.0869,-85.9379 71.4062,-93.5765\"/>\n",
       "</g>\n",
       "<!-- anon_1_3_1&#45;&#45;3 -->\n",
       "<g id=\"edge14\" class=\"edge\"><title>anon_1_3_1&#45;&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M68.1336,-97.4628C61.4787,-101.284 42.3458,-112.269 28.932,-119.97\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7fc2166649b0>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h, network_method='msn', debug=False, show_node_labels=True, node_size_factor=.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"164pt\" height=\"229pt\"\n",
       " viewBox=\"0.00 0.00 164.39 229.11\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(76.5742 102.702)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-76.5742,126.41 -76.5742,-102.702 87.8128,-102.702 87.8128,126.41 -76.5742,126.41\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>0</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-49.5742\" cy=\"-14.0886\" rx=\"23\" ry=\"23\"/>\n",
       "<text text-anchor=\"middle\" x=\"-49.5742\" y=\"-10.3886\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 4 -->\n",
       "<g id=\"node5\" class=\"node\"><title>4</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"41.1536\" cy=\"33.9573\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"41.1536\" y=\"37.6573\" font-family=\"Times,serif\" font-size=\"14.00\">4</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;4 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>0&#45;&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-28.9866,-3.18626C-5.98245,8.99588 29.2837,27.6715 38.7473,32.6831\"/>\n",
       "</g>\n",
       "<!-- 5 -->\n",
       "<g id=\"node6\" class=\"node\"><title>5</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"15.361\" cy=\"-58.0153\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"15.361\" y=\"-54.3153\" font-family=\"Times,serif\" font-size=\"14.00\">5</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;5 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>0&#45;&#45;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-30.4427,-27.0305C-14.5806,-37.7607 6.42164,-51.9681 13.1726,-56.5349\"/>\n",
       "</g>\n",
       "<!-- 6 -->\n",
       "<g id=\"node7\" class=\"node\"><title>6</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-31.1696\" cy=\"19.3479\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"-31.1696\" y=\"23.0479\" font-family=\"Times,serif\" font-size=\"14.00\">6</text>\n",
       "</g>\n",
       "<!-- 0&#45;&#45;6 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>0&#45;&#45;6</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-38.4436,6.13289C-36.0211,10.5339 -33.7853,14.5957 -32.4492,17.0233\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>1</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"67.8128\" cy=\"-22.1213\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"67.8128\" y=\"-18.4213\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;4 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>1&#45;&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M60.8084,-7.38729C54.3908,6.11251 45.4266,24.9691 42.2906,31.5657\"/>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;5 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>1&#45;&#45;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M54.577,-31.1789C41.952,-39.8184 23.9539,-52.1349 17.6485,-56.4499\"/>\n",
       "</g>\n",
       "<!-- 7 -->\n",
       "<g id=\"node8\" class=\"node\"><title>7</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"18.2736\" cy=\"0.702473\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"18.2736\" y=\"4.40247\" font-family=\"Times,serif\" font-size=\"14.00\">7</text>\n",
       "</g>\n",
       "<!-- 1&#45;&#45;7 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>1&#45;&#45;7</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M53.2174,-15.3969C41.5727,-10.0319 26.2971,-2.99413 20.5718,-0.356371\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\"><title>2</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"2.89261\" cy=\"106.41\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"2.89261\" y=\"110.11\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 8 -->\n",
       "<g id=\"node9\" class=\"node\"><title>8</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"3.58406\" cy=\"46.5568\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"3.58406\" y=\"50.2568\" font-family=\"Times,serif\" font-size=\"14.00\">8</text>\n",
       "</g>\n",
       "<!-- 2&#45;&#45;8 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>2&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M3.07791,90.3703C3.24432,75.9653 3.47445,56.0447 3.5549,49.0808\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\"><title>3</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"-53.6022\" cy=\"-82.7024\" rx=\"16\" ry=\"16\"/>\n",
       "<text text-anchor=\"middle\" x=\"-53.6022\" y=\"-79.0024\" font-family=\"Times,serif\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 9 -->\n",
       "<g id=\"node10\" class=\"node\"><title>9</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"-14.7317\" cy=\"-30.0467\" rx=\"2.5\" ry=\"2.5\"/>\n",
       "<text text-anchor=\"middle\" x=\"-14.7317\" y=\"-26.3467\" font-family=\"Times,serif\" font-size=\"14.00\">9</text>\n",
       "</g>\n",
       "<!-- 3&#45;&#45;9 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>3&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M-43.9938,-69.6864C-34.6413,-57.0172 -21.1691,-38.767 -16.4458,-32.3686\"/>\n",
       "</g>\n",
       "<!-- 4&#45;&#45;8 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>4&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M38.6886,34.784C31.8975,37.0615 12.9675,43.4099 6.10744,45.7105\"/>\n",
       "</g>\n",
       "<!-- 5&#45;&#45;9 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>5&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M13.3865,-56.1802C7.94697,-51.1246 -7.21573,-37.0321 -12.7105,-31.9252\"/>\n",
       "</g>\n",
       "<!-- 6&#45;&#45;8 -->\n",
       "<g id=\"edge11\" class=\"edge\"><title>6&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M-28.8893,21.1331C-22.4865,26.146 -4.41838,40.2916 1.60337,45.0061\"/>\n",
       "</g>\n",
       "<!-- 6&#45;&#45;9 -->\n",
       "<g id=\"edge12\" class=\"edge\"><title>6&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M-30.3768,16.9655C-27.6063,8.64042 -18.373,-19.1047 -15.5553,-27.5718\"/>\n",
       "</g>\n",
       "<!-- 7&#45;&#45;8 -->\n",
       "<g id=\"edge13\" class=\"edge\"><title>7&#45;&#45;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M17.4838,3.1678C14.9571,11.0552 7.08292,35.6349 4.45082,43.8511\"/>\n",
       "</g>\n",
       "<!-- 7&#45;&#45;9 -->\n",
       "<g id=\"edge14\" class=\"edge\"><title>7&#45;&#45;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-dasharray=\"5,2\" d=\"M16.108,-1.31507C10.0273,-6.98015 -7.13184,-22.9663 -12.8507,-28.2942\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Graph at 0x7fc2166646d8>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_haplotype_network(h[:, :], network_method='mjn', debug=False, show_node_labels=True, node_size_factor=.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
